---
title: "Introduction to imaging-based spatial transcriptomics analysis (10x Xenium breast cancer)"
date: "`r format(Sys.time(), '%b %Y')`"
title-block-banner: true
author: 
  - name: Dharmesh D. Bhuva
    orcid: 0000-0002-6398-9157
    url: https://bhuvad.github.io/
    email: dharmesh.bhuva@adelaide.edu.au
    affiliation:
      - SAiGENCI, The University of Adelaide
      - South Australian Health and Medical Research Institute (SAHMRI)
      - Walter and Eliza Hall Institute
  - name: Chin Wee Tan
    orcid: 0000-0001-9695-7218
    email: cwtan@wehi.edu.au
    affiliation:
      - Walter and Eliza Hall Institute
      - The University of Melbourne
output:
  prettydoc::html_pretty:
    theme: cayman
    toc: yes
    toc_depth: 2
    number_sections: yes
    fig_caption: yes
    df_print: paged
comments:
  hypothesis: 
    theme: clean
editor: source
nocite: | 
  @R-ggplot2, @R-knitr, @R-rmarkdown, @R-prettydoc
bibliography: [bibliography.bib, packages.bib]
vignette: >
  %\VignetteIndexEntry{Introduction to imaging-based spatial transcriptomics analysis (10x Xenium breast cancer).}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
keywords: spatial transcriptomics, cell typing, normalisation, signature scoring, visualisation
---

<p>

**R version**: `r R.version.string` <br /> **Bioconductor version**: `r BiocManager::version()` <br /> **Package version**: `r packageVersion("XeniumSpatialAnalysisWorkshop")`

</p>

```{r setup, include=FALSE}
#set knitr chunk options
knitr::opts_chunk$set(warning = FALSE, message = FALSE)

#load packages to avoid startup messages later in the code
library(ggplot2)
library(patchwork)
library(scran)
library(scater)
library(standR)
library(SingleR)
library(SpatialExperiment)
library(SpaNorm)
library(singscore)
library(msigdb)
library(vissE)
library(igraph)
library(XeniumSpatialAnalysisWorkshop)

#automatically create a bib database for R packages
knitr::write_bib(c(
  .packages(), 'knitr', 'rmarkdown', 'prettydoc'
), 'packages.bib')
```

# Introduction

Spatial molecular technologies have revolutionised the study of biological systems. We can now study cellular and sub-cellular structures in complex systems. However, these technologies generate complex datasets that require new computational approaches and ideas to analyse. Many of these pipelines are still being developed therefore what is state-of-the-art now will not remain so for too long. However, the biological aim of analysis will mostly remain the same with more being uncovered as computational methods advance. This workshop intends to demonstrate the types of biological questions that can be studied using spatial transcriptomics datasets. Our intention is NOT to use the state-of-the-art methods, but to show you how data can be analysed to study spatial hypotheses. As users get familiar with these ideologies, they can swap out older tools with newer, more powerful tools. As instructors, we hope to update this workshop as more and better tools become available. We look forward to receiving community contributions as well.

# Pre-processing steps already performed

Data from the instrument comes in raw images that are often summerised to transcript tables. These tables are then coupled with immunofluorescence images for a few key proteins or marker such as DAPI which stains the nucleus, Pan-Cytokeratin which stains the membrane of cancer cells, a cocktail of membrane proteins, markers such as CD45 which stain immune cells, and a cocktail of cytoplasmic markers (technology specific). These proteins are used to perform cell segmentation as the measurements are obtained at the transcript level and not the cell level (as is the case with single-cell RNA-seq data). Platforms provide segmented data and users can opt to perform their own segmentation if needed. If you wish you re-segment the data yourself, you could try using methods such as BIDCell [@fu2024bidcell] that use a single cell reference dataset to perform segmentation using the marker proteins as well as transcriptomics measurements. Do note that vendors are constantly improving segmentation and recent improvements are accurate on difficult to segment cells such as neurons and astrocytes.

We will use the already segmented cell-level count data. We will also use a previously prepared SpatialExperiment object. If users want to prepare their own object, all they need is the cell matrix file, the feature table, the cell table and a table of spatial coordinates (if this information is not already in the cell table). With these, information on preparing your own SpatialExperiment object can be found in this package's [vignette](https://www.bioconductor.org/packages/release/bioc/vignettes/SpatialExperiment/inst/doc/SpatialExperiment.html). The design for this object is shown in the figure below.

![SpatialExperiment object schematic](https://raw.githubusercontent.com/drighelli/SpatialExperiment/refs/heads/devel/vignettes/SPE.png)

# Loading prepared data

The data we will use in this workshop is from the 10x Xenium platform. Since the objective of this workshop is to introduce you to spatial transcriptomics analysis, we do not work with the full dataset, which contains over 500,000 cells and instead opt to work with a smaller subset of the sample. This region contains interesting features and is therefore enough to showcase a complete spatial transcriptomics analysis workflow. The data used in this workshop is a window with the x coordinate in the interval (4800, 5800) and the y coordinate in the interval (8500, 9500). This window is shown in the context of the full tissue below.

![Full IDC](https://raw.githubusercontent.com/bhuvad/XeniumSpatialAnalysisWorkshop/refs/heads/main/inst/extdata/full_tissue.png)

```{r}
data(idc)
# preview object
idc
# retrieve column (cell) annotations
colData(idc)
# retrieve row (gene/probe) annotations
rowData(idc)
# retrieve spatial coordinates
head(spatialCoords(idc))
# retrieve counts - top 5 cells and genes
counts(idc)[1:5, 1:5]
```

```{r}
# set default point size to 0.5 for this report
update_geom_defaults(geom = "point", new = list(size = 0.5))
plotSpatial(idc, colour = region) +
  scale_colour_brewer(palette = "Dark2") +
  # improve legend visibility
  guides(colour = guide_legend(override.aes = list(shape = 15, size = 5)))
```

For users interested in analysing the whole dataset, it can be obtained using the code below.

> Please note that the full dataset contains over 500,000 cells and will not scale for the purpose of this workshop. If interested, run the code below on your own laptop.

```{r eval = FALSE}
## DO NOT RUN during the workshop as the data are large and could crash your session
if (!require("ExperimentHub", quietly = TRUE)) {
  BiocManager::install("ExperimentHub")
}
if (!require("SubcellularSpatialData", quietly = TRUE)) {
  BiocManager::install("SubcellularSpatialData")
}

# load the metadata from the ExperimentHub
eh = ExperimentHub()
# download the transcript-level data
tx = eh[["EH8567"]]
# filter the IDC sample
tx = tx[tx$sample_id == "IDC", ]
# summarise measurements for each cell
idc = tx2spe(tx)
```

# Quality control

Most spatial transcriptomics assays include negative control probes to aid in quality control. These probes do not bind to the transcripts of interest (signal), so represent noise in the assay. There are three different types of negative controls in the 10x Xenium assay. As these vary across technologies and serve different purposes, we treat them equally for our analysis and assess them collectively. The breakdown of the different types of negative control probes is given below.

```{r}
table(rowData(idc)$genetype)
```

We now use compute and visualise quality control metrics for each cell. Some of the key metrics we assess are:

1. `sum` or `total` - Total number of transcripts and negative control probes per cell (more commonly called the library size).
1. `detected` - Number of transcripts and negative control probes detected per cell (count > 0).
1. `subsets_neg_sum` - Count of negative control probes (noise) per cell.
1. `subsets_neg_detected` - Number of negative control probes expressed per cell (count > 0).
1. `subsets_neg_percent` - `subsets_neg_sum` / `sum`.

Additionally, the vendor (10x) provides their own summary of quality through the `qv` score. This is usually provided per detected measurements but we compute the average per cell when deriving cellular summaries. Likewise, we compute the number of transcripts allocated to the nucleus (`overlaps_nucleus`).

```{r}
# identify the negative control probes
is_neg = rowData(idc)$genetype != "Gene"
# compute QC metrics for each cell
idc = addPerCellQCMetrics(idc, subsets = list("neg" = is_neg))
```

> Quality control is all about visualisation!

One of the best ways to understand the data and identify problematic cells or regions is to visualise the QC metrics. We need to visualise the metrics in various contexts to better understand what cells appear of poor qaulity. As such, we usually plot metrics against each other, we visualise them in space, and we visualise them in the context of the higher dimensional structure of the data. To aid the latter, we will compute lower dimensional representations using the t-distributed stochastic neighbours embedding (t-SNE). As the t-SNE embedding can be computationally intensive, we usually perform PCA prior to t-SNE. This approach to quality control is identical to that used for single-cell RNA-seq analysis.

```{r}
# compute log counts - to fix the skew with counts
logcounts(idc) = log2(counts(idc) + 1)
# compute PCA
idc = runPCA(idc)
# compute t-SNE
idc = runTSNE(idc, dimred = "PCA")
```

We plot the `sum` (library size) against the number of genes detected, and colouring the points based on the quality score (`qv`). We see that as more genes are detected, more measurements are obtained (library size increases). There are cells where the library size is very low and the number of genes detected is low as well. Considering the smaller panel size of 380 target genes in this assay, we expect the number of genes detected in some cell types to be small. However, if we detect very few genes in a cell, we are less likely to obtain any useful information from them. As such, poor quality cells can be defined as those with very low library sizes and number of genes detected.

```{r}
plotColData(idc, "detected", "sum", colour_by = "qv") +
  geom_hline(yintercept = 20, lty = 2) +
  geom_vline(xintercept = 10, lty = 2)
```

However, we need to identify appropriate thresholds to classify poor quality cells. As the panel sizes and compositions for these emerging technologies are variable, there are no standards as of yet for filtering. For instance, immune cells will always have smaller library sizes and potentially a smaller number of genes expressed in a targetted panel as each cell type may be represented by a few marker genes. Additionally, these cells are smaller in size therefore could be filtered out using an area-based filter. Desipte the lack of standards, the extreme cases where cells express fewer than 10 genes and have fewer than 10-20 detections are clearly of poor quality. As such, we can start with these thresholds and adjust based on other metrics and visualisations.

```{r}
# t-SNE plot
p1 = plotDR(idc, dimred = "TSNE", colour = sum, alpha = 0.75, size = 0.5) +
  scale_colour_viridis_c(option = "F") +
  labs(x = "TSNE1", y = "TSNE2") +
  theme(legend.position = "bottom")
# spatial plot
p2 = plotSpatial(idc, colour = sum, alpha = 0.75) +
  scale_colour_viridis_c(option = "F") +
  theme(legend.position = "bottom")
p1 + p2 + plot_annotation(title = "Sum (library size)")
```

```{r}
# t-SNE plot
p1 = plotDR(idc, dimred = "TSNE", colour = detected, alpha = 0.75, size = 0.5) +
  scale_colour_viridis_c(option = "F") +
  labs(x = "TSNE1", y = "TSNE2") +
  theme(legend.position = "bottom")
# spatial plot
p2 = plotSpatial(idc, colour = detected, alpha = 0.75) +
  scale_colour_viridis_c(option = "F") +
  theme(legend.position = "bottom")
p1 + p2 + plot_annotation(title = "Detected (number of genes expressed)")
```

```{r}
# t-SNE plot
p1 = plotDR(idc, dimred = "TSNE", colour = qv, alpha = 0.75, size = 0.5) +
  scale_colour_viridis_c(option = "F") +
  labs(x = "TSNE1", y = "TSNE2") +
  theme(legend.position = "bottom")
# spatial plot
p2 = plotSpatial(idc, colour = qv, alpha = 0.75) +
  scale_colour_viridis_c(option = "F") +
  theme(legend.position = "bottom")
p1 + p2 + plot_annotation(title = "Quality score")
```

```{r}
# t-SNE plot
p1 = plotDR(idc, dimred = "TSNE", colour = subsets_neg_percent, alpha = 0.75, size = 0.5) +
  scale_colour_viridis_c(option = "F", limits = c(0, 2), oob = scales::squish) +
  labs(x = "TSNE1", y = "TSNE2") +
  theme(legend.position = "bottom")
# spatial plot
p2 = plotSpatial(idc, colour = subsets_neg_percent, alpha = 0.75) +
  scale_colour_viridis_c(option = "F", limits = c(0, 2), oob = scales::squish) +
  theme(legend.position = "bottom")
p1 + p2 + plot_annotation(title = "Proportion of negative control probes")
```

```{r}
# t-SNE plot
p1 = plotDR(idc, dimred = "TSNE", colour = overlaps_nucleus, alpha = 0.75, size = 0.5) +
  scale_colour_viridis_c(option = "F") +
  labs(x = "TSNE1", y = "TSNE2") +
  theme(legend.position = "bottom")
# spatial plot
p2 = plotSpatial(idc, colour = overlaps_nucleus, alpha = 0.75) +
  scale_colour_viridis_c(option = "F") +
  theme(legend.position = "bottom")
p1 + p2 + plot_annotation(title = "Proportion of nuclear transcripts")
```

Having visualised the quality metrics and their distrition across space and the genomic dimensions, we can now begin to test threshold values. We want to remove low quality cells but we have to be careful not to introduce bias (e.g., by preferentially removing biological entities such as regions or cells).

Try modifying the thresholds below to see the effect of over- or under- filtering.

* What happens when you set the `sum` threshold to 100?
* What about setting the `detected` threshold to 50?

```{r}
thresh_sum = 20
thresh_detected = 10

# classify cells to keep
idc$keep = idc$sum > thresh_sum & idc$detected > thresh_detected

# t-SNE plot
p1 = plotDR(idc, dimred = "TSNE", colour = keep, alpha = 0.75, size = 0.5) +
  scale_color_brewer(palette = "Set1") +
  labs(x = "TSNE1", y = "TSNE2") +
  theme(legend.position = "bottom") +
  guides(colour = guide_legend(override.aes = list(shape = 15, size = 5)))
# spatial plot
p2 = plotSpatial(idc, colour = keep, alpha = 0.75) +
  scale_color_brewer(palette = "Set1") +
  theme(legend.position = "bottom") +
  guides(colour = guide_legend(override.aes = list(shape = 15, size = 5)))
p1 + p2 + plot_annotation(title = "Cells retained after QC filtering")
```

In practice, after poor quality cells are filtered out, we tend to see clearer separation of clusters in the t-SNE (or equivalent visualisations). We also notice that poor quality cells often seem like intermmediate states between clusters in these plots. Additionally, they tend to be randomly distributed in space. If they are concentrated in some sub-regions (as above - top left corner), it is usually due to technical reasons. In the above plot, we see a clear rectangular section that is of poor quality. It is unlikely that biology would be present in such a regular shape therefore we can use that region as a guide in selecting thresholds. Having selected the thresholds, we now apply them and remove the negative control probes from donwstream analysis.

```{r}
# apply the filtering
idc = idc[, idc$keep]
# remove negative probes
idc = idc[!is_neg, ]
```

Newer quality control tools are being developed and are currently under peer-review. These include SpotSweeper [@totty2024spotsweeper] and [SpaceTrooper](https://github.com/drighelli/SpaceTrooper). As these software are not deposited to more permanent repositories, we do not use them in this workshop. We advise advanced users to use these tools for their QC as they help come up with additional QC summaries for each cell or region in space. SpaceTrooper for instance, computes a flag score that penalises cells that are too large or small, and whose aspect ratio is large. While these filters may make sense in most systems, they could penalise stomal or neuronal cells that are elongated and whose aspect ratio will be very large. Therefore always remember that QC is relative the biological system being studied.

As these technologies are still emerging, all quality control metrics should be evaluated on a case-by-case basis depending on the platform, panel set, and tissue. Knowing the biology to expect or what is definitely NOT biology helps in making these decisions. As any experienced bioinformatician will tell you, you will vary these thresholds a few times throughout your analysis because you learn more about the data, biology, and platform as you analyse the data. This cyclical approach can introduce biases so always rely on multiple avenues of evidence before changing parameters.

# Normalisation

Above we plot the `sum`, which is commonly known as the library size, for the purpose of quality control. Assessing the library size is helpful in identifying poor quality cells. Beyond QC, variation in library size can affect donwstream analysis as it represents measurement bias. A gene may appear to be highly expressed in some cells simply because more measurements were sampled from the cell, and not because the gene is truly highly expressed. Whether it is sequencing-based or imaging-based transcriptomics, we are always sampling from the true pool of transcripts and are therefore likely to have library size variation.

```{r}
idc$logLS = log2(idc$sum + 1)
plotSpatial(idc, colour = logLS) +
  scale_colour_viridis_c(option = "F") +
  theme(legend.position = "bottom")
```

To remove these library size biases, we perform library size normalisation. As measurements are independent in bulk and single-cell RNA-seq (reactions happen independently for each sample/cell), each cell or sample can be normalised independently by computing sample/cell-specific adjustment factors. Measurements are locally dependent in spatial transcriptomics datasets therefore library size effects confound biology [@bhuva2024library]. The plot above show this clearly as the technical effect of library size is already showing tissue structures. These can be due to various causes such as:

1. The tissue structure in one region is more rigid, thereby making it difficult for reagents to permeate.
1. Reagents distributed unevenly due to technical issues leading to a sub-region within a tissue region being under sampled.

In both cases, each cell cannot be normalised independently as nearby cells are likely to experience similar technical effects. To tackle this, our team has developed the SpaNorm normalisation method that uses spatial dependence to decouple technical and biological effects. SpaNorm uses a spatially constrained regularised generalised linear model to model library size effects and subsequently adjust the data.

![SpaNorm schematic](https://raw.githubusercontent.com/bhuvad/SpaNorm/refs/heads/master/vignettes/SpaNormWorkflow.png)

The key parameters for SpaNorm normalisation that need to be adjusted are:

1. `sample.p` - This is the proportion of the data that is sub-sampled to fit the regularised GLM. As spatial transcriptomics datasets are very large, the model can be approximated using by sampling. In practice, we see that ~50,000 cells in a dataset with 500,000 cells works well enough under the assumption that all possible cellular states are sampled evenly across the tissue. This is likely to be true in most cases. Below we use half the cells to fit the model (you can tweak this to 1 to use all cells).
1. `df.tps` - Spatial dependence is modelled using thin plate spline functions. This parameter controls the degrees of freedom (complexity) of the function. Complex functions will fit the data better but could overfit therefore in practice, we see that 6-10 degrees of freedom are generally good enough. If the degrees of freedom are increased, regularisation should also be increased (`lambda.a` parameter).

```{r}
set.seed(1000)
# perform SpaNorm normalisation
idc = SpaNorm(idc, sample.p = 0.5, df.tps = 6)
```

Above, the model is fit and the default adjustment, `logPAC`, is computed. Alternative adjustments such as the mean biology (`meanbio`), pearson residuals (`pearson`), and median biology (`median`) can be computed as well (this will be faster as the fit does not need to be repeated).

```{r}
# compute mean biology
idc_mean = SpaNorm(idc, adj.method = "meanbio")

p1 = plotSpatial(idc, what = "expression", assay = "logcounts", colour = EPCAM, alpha = 0.75) +
  scale_colour_viridis_c(option = "F") +
  ggtitle("Normalised expression") +
  theme(legend.position = "bottom")
p2 = plotSpatial(idc_mean, what = "expression", assay = "logcounts", colour = EPCAM, alpha = 0.75) +
  scale_colour_viridis_c(option = "F") +
  ggtitle("Mean from SpaNorm model") +
  theme(legend.position = "bottom")
p1 + p2 + plot_annotation(title = "Proportion of nuclear transcripts")
```

[_EPCAM_](https://www.genecards.org/cgi-bin/carddisp.pl?gene=DHRS) is an epithelial cell marker gene that will also be expressed in breast cancer cells that are of epithelial origin. The SpaNorm model correctly models the regions where the cancer cells exist (mean) and is able to normalise technical effects without affecting the biology.

Library size normalisation is important to ensure technical variation is removed from gene expression. In the most extreme cases, a gene that is highly expressed in a region of interest may seem to be lowly expressed because of a drop in library sizes in said region (see manuscript). We have seen that negative probes across most technologies are also capturing library size effects very well and are experimenting with using these to model technical variation and normalise data. In some technologies/datasets, these are very lowly expressed, not because of better data quality but by design as well, and therefore would NOT faithfully represent technical variation. As with every other step, this is still an active area of research.

# Cell typing

Cell typing can be performed using various approaches. Most of these were developed for the analysis of single-cell RNA-seq datasets and port across well to spatial transcriptomics datasets. The two common strategies are reference-based and reference-free annotations. In the former, annotations are mapped across from a previously annotated dataset. For a thorough discussion on this topic, review the [Orchestrating Single-Cell Analysis (OSCA) with Bioconductor](https://bioconductor.org/books/release/OSCA/) book [@amezquita2020orchestrating]. Here, we use the SingleR package [@aran2019reference] along with a reference breast cancer dataset [@wu2021single]. The processed dataset can be downloaded from [Curated Cancer Cell Atlas](https://www.weizmann.ac.il/sites/3CA/breast). 

SingleR is a simple yet powerful algorithm that only requires a single prototype of what the cell type's expression profile should look like. As such, counts are aggregated (pseudobulked) across cells to form a reference dataset where a single expression vector is created for each cell type. Aggregation is performed using the `scater::aggregateAcrossCells` function. As the original dataset contains >96,000 cells, we only provide the aggregated counts for this workshop.

```{r}
# load reference dataset - Wu et al., Nat Genetics, 2021
data(ref_wu)
ref_wu
# cell types in the reference dataset
ref_wu$label.cell_type
```

```{r}
# predict cell types using SingleR
prediction = SingleR(test = idc, ref = ref_wu, labels = ref_wu$label.cell_type)
# add prediction to the SpatialExperiment object
idc$cell_type = prediction$labels
```

```{r}
plotSpatial(idc, colour = cell_type, alpha = 0.75) +
  scale_colour_brewer(palette = "Paired", na.value = "#333333") +
  ggtitle("Mean from SpaNorm model") +
  theme(legend.position = "bottom") +
  guides(colour = guide_legend(override.aes = list(shape = 15, size = 5)))
```

# Spatial domain identification

Regions were annotated by an expert pathologist in our dataset. However, this is not always feasible, either due to access to a pathologist, or because the features are not visible in histology images alone. In such instances, a computational approach is required to identify spatial domains or cellular niches. These domains are generally, but not always, defined based on the homogeneity of biology. As this is still a developing area, there are many algorithms available, however, they generally fall within these three categories:

1. Graph Neural Network (GNN) or Graph Convolutional Network (GCN) based algorithms.
1. Probabilistic models based on markov random fields.
1. Neighbourhood diffusion followed by graph-based clustering.

We note that this is not a comprehensive survey of the literature but just a summary of the most popular methods in this space. The first two approaches can be very powerful, however, are time consuming and will not always scale with larger datasets. Diffusion followed by graph-clustering is a relatively simple, yet powerful approach that scales to large datasets. Banksy is one such algorithm that first shares information within a local neighbourhood, and then performs a standard single-cell RNA-seq inspired graph-based clustering to identify spatial domains [@singhal2024banksy].

The following key parameters determine the final clustering results:

1. `lambda` - This controls the degree of information sharing in the neighbourhood (range of [0, 1]) where larger values result in higher diffusion and thereby smoother clusters.
1. `resolution` - The resolution parameter for the graph-based clustering using either the Louvain or Leiden algorithms. Likewise, there are other parameters to control the graph-based clustering. The OSCA book discusses choices for these parameters in depth.
1. `npcs` - The number of principal components to use to condense the feature space. The OSCA book discusses choices for this in depth.
1. `k_geom` - The number of nearest neighbours used to share information. The authors recommend values within [15, 30] and note that variation is minimal within this range.

```{r}
# parameters
lambda = 0.2
npcs = 50
k_geom = 30
res = 0.8

# compute the features required by Banksy
idc = Banksy::computeBanksy(idc, assay_name = "logcounts", k_geom = k_geom)

# compute PCA on the features
set.seed(1000)
idc = Banksy::runBanksyPCA(idc, lambda = lambda, npcs = npcs)

# perform graph-based clusering
set.seed(1000)
idc = Banksy::clusterBanksy(idc, lambda = lambda, npcs = npcs, resolution = res)

# store to the column predicted_region
idc$predicted_region = idc$`clust_M0_lam0.2_k50_res0.8`

plotSpatial(idc, colour = predicted_region, alpha = 0.75) +
  scale_colour_brewer(palette = "Paired", na.value = "#333333", name = "cluster") +
  ggtitle("Mean from SpaNorm model") +
  theme(legend.position = "bottom") +
  guides(colour = guide_legend(override.aes = list(shape = 15, size = 5)))
```

Try varying the parameters to see how the clustering changes. Do note that the results are automatically stored in a new column in the colData. Unfortunately, the function does not allow specification of the new column name so the only approach is to look at the column names using `colnames(colData(idc))` and find the name of the new column (usually the last one) containing the clustering results. We then store these to a column of our choosing to make it easier to reference it.

Looking back at the original annotations, we see that clusters 3 and 5 represent the DCIS and Invasive tumours respectively. Similarly, cluster 10 represents the blood vessel annotated by the pathologist. Computational analysis reveals additional structures present in the tissue that may differ only in the genomic measurements. Clustes 6 and 7 for instance seems to represent different structures in the stroma, some of which may be visible in the histology image but are intestive to label manually. We also find cluster 4 which represents a boundary of cells surrounding invasive and DCIS lesions.

```{r}
p1 = plotSpatial(idc, colour = predicted_region, alpha = 0.75) +
  scale_colour_brewer(palette = "Paired", na.value = "#333333", name = "cluster") +
  ggtitle("Mean from SpaNorm model") +
  theme(legend.position = "bottom") +
  guides(colour = guide_legend(nrow = 3, override.aes = list(shape = 15, size = 5)))
p2 = plotSpatial(idc, colour = region) +
  scale_colour_brewer(palette = "Dark2") +
  ggtitle("Pathology annotations") +
  guides(colour = guide_legend(nrow = 3, override.aes = list(shape = 15, size = 5)))
p1 + p2 + plot_annotation(title = "Spatial domain identification")
```

# Differential expression

Cell type identification or clustering allows us to identify the different cellular states in a biological system. The power of spatial molecular datasets is that we can study how cellular states change in response to different structures in the tissue. For instance, in our tissue we see that there are two different types of cancer lesions, ductal carcinoma in situ and invasive. Cancer cells in the latter lesion tend to me aggresive and can spread more than those in the DCIS region. We would be interested in knowning the biological processes that make this happen.

Additionally, cancer cells interact with their environment including cell types such as fibroblasts. We may be interested in identifying the differences in fibroblasts that are in close proximity to cancer cells as opposed to the distant ones. We can identify the genes that are associated with these changes below.

```{r}
# create clusters to study how cell types vary across domains
clusters = paste(idc$cell_type, idc$predicted_region, sep = "_")
clusters |>
  table() |>
  sort() |> 
  tail(10)
```

We see that the two largest clusters capture malignant cells in spatial domains 2 and 5 which align with the invasive and DCIS regions respectively based on the pathology annotations. We also see that though most of the fibroblasts are dispersed in the stroma (region 6), those that are closer to the cancer lesions are clustered in region 3. We would be interested in studying the difference between these and understanding whether they contain features that promote the tumour.

> The analysis below is only meant to be a demonstration of the potential of spatial transcriptomics data. A proper statistical analysis would require replicate tissue samples from across different patients. In such a case, we would aggregate transcript counts across cell types using the `scater::sumCountsAcrossCells` to create pseudo-bulk samples, and follow up with a differential expression analysis using the [`edgeR`](https://doi.org/10.12688/f1000research.8987.2) or [`limma`](https://doi.org/10.12688/f1000research.9005.3) pipelines. Alternatively, a more advanced tool such as [`nicheDE`](https://doi.org/10.1186/s13059-023-03159-6) could be used to perform analysis on a single tissue, however, it would also be more robust with biological replicates. This tool is outside the scope of this introductory workshop therefore is not covered.

```{r}
library(scran)

# identify markers
marker.info = scoreMarkers(idc, clusters, pairings = list(c("Malignant_2", "Fibroblast_3"), c("Malignant_5", "Fibroblast_6")))
marker.info
```

Below we compare the malignant (cancer) cells from region 2 (invasive) against those from region 5 (DCIS). Investigating and visialising the top genes shows that [_DHRS_](https://www.genecards.org/cgi-bin/carddisp.pl?gene=DHRS) and [_MUC6_](https://www.genecards.org/cgi-bin/carddisp.pl?gene=MUC6) are up-regulated in invasive cancer cells while [_GATA3_](https://www.genecards.org/cgi-bin/carddisp.pl?gene=GATA3) and [_FLNB_](https://www.genecards.org/cgi-bin/carddisp.pl?gene=FLNB) are moderately up-regulated in DCIS cancer cells.

```{r fig.width=8, fig.height=8}
library(scater)

# study markers of Malignant_2
chosen_malignant <- marker.info[["Malignant_2"]]
ordered_malignant <- chosen_malignant[order(chosen_malignant$rank.AUC), ]
head(ordered_malignant[, 1:4])

plotExpression(idc,
  features = head(rownames(ordered_malignant), 9),
  x = "predicted_region",
  colour_by = "predicted_region",
  ncol = 3
) +
  geom_vline(xintercept = c(2, 5), lwd = 0.1, lty = 2) +
  scale_colour_brewer(palette = "Paired", na.value = "#333333", name = "cluster") +
  theme(legend.position = "bottom") +
  guides(colour = guide_legend(override.aes = list(shape = 15, size = 5)))
```

Next we compare the fibroblasts from region 6 (stroma) against those from region 3 (tumour-adjacent). Investigating and visialising the top genes shows that [_LUM_](https://www.genecards.org/cgi-bin/carddisp.pl?gene=LUM) and [_POSTN_](https://www.genecards.org/cgi-bin/carddisp.pl?gene=POSTN) are up-regulated in stromal fibroblasts while [_MMP11_](https://www.genecards.org/cgi-bin/carddisp.pl?gene=MMP11) and [_AQP1_](https://www.genecards.org/cgi-bin/carddisp.pl?gene=AQP1) are up-regulated in the tumour-adjacent fibroblasts.

```{r fig.width=8, fig.height=8}
# study markers of Malignant_2
chosen_fibroblasts <- marker.info[["Fibroblast_3"]]
ordered_fibroblasts <- chosen_fibroblasts[order(chosen_fibroblasts$rank.AUC), ]
head(ordered_fibroblasts[, 1:4])

plotExpression(idc,
  features = head(rownames(ordered_fibroblasts), 9),
  x = "predicted_region",
  colour_by = "predicted_region",
  ncol = 3
) +
  geom_vline(xintercept = c(3, 6), lwd = 0.1, lty = 2) +
  scale_colour_brewer(palette = "Paired", na.value = "#333333", name = "cluster") +
  theme(legend.position = "bottom") +
  guides(colour = guide_legend(override.aes = list(shape = 15, size = 5)))
```

# Functional analysis

The example dataset used here was generated using the 10x Xenium platform that measured 380 genes. More recent versions of this platform and competing platforms such as the NanoString CosMx are now providing around 5000 genes (soon whole transcriptome as well!). It is relatively easy to scan through the strongest differentially expressed genes from a list of at most 380 genes to understand the processes underpinning the system. With larger panel sizes, we need computational tools to perform such analyses, often referred to as gene-set enrichment analysis or functional analysis. As this is an extensive topic, it is not covered in depth in this workshop, however, we have previously developed a [workshop on gene-set enrichment analysis](https://davislaboratory.github.io/GenesetAnalysisWorkflow/) and refer the attendees to that if they are interested. A non-programmatic interface of some of the tools in our workshop can be accessed through [vissE.cloud](https://visse.cloud/).

> Optional: You can also explore the full dataset from this workshop on vissE.cloud through the following [link](https://visse.cloud/galleryexample/xenium_V1FFPEHumanBreastIDCWithAddon_PCA). This exploratory analysis performs principal components analysis (PCA) or non-negative matrix factorisation (NMF) on the dataset to identify macroscopic biological programs. It then performs gene-set enrichment analysis on the factor loadings to explain the biological processes.

# Packages used {.unnumbered}

This workflow depends on various packages from version `r BiocManager::version()` of the Bioconductor project, running on `r version$version.string` or higher. The complete list of the packages used for this workflow are shown below:

```{r session_info}
sessionInfo()
```

# Acknowledgments {.unnumbered}


# References {.unnumbered}
