<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en-US">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to imaging-based spatial transcriptomics analysis (Bruker CosMx NSCLC) • CosMxSpatialAnalysisWorkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to imaging-based spatial transcriptomics analysis (Bruker CosMx NSCLC)">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CosMxSpatialAnalysisWorkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/workflow_spatial_cosmx.html">Introduction to imaging-based spatial transcriptomics analysis (Bruker CosMx NSCLC)</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/bhuvad/XeniumSpatialAnalysisWorkshop">
    <span class="fa fa-github"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to imaging-based spatial
transcriptomics analysis (Bruker CosMx NSCLC)</h1>
                        <h4 data-toc-skip class="author">Dharmesh D.
Bhuva</h4>
            <address class="author_afil">
      Frazer Institute, The University of QueenslandSAiGENCI, The
University of
Adelaide<br><a class="author_email" href="mailto:#"></a><a href="mailto:dharmesh.bhuva@adelaide.edu.au" class="email">dharmesh.bhuva@adelaide.edu.au</a>
      </address>
                              <h4 data-toc-skip class="author">Chin Wee
Tan</h4>
            <address class="author_afil">
      Walter and Eliza Hall InstituteThe University of
Melbourne<br><a class="author_email" href="mailto:#"></a><a href="mailto:cwtan@wehi.edu.au" class="email">cwtan@wehi.edu.au</a>
      </address>
                  
            <h4 data-toc-skip class="date">Jul 2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/bhuvad/CosMxSpatialAnalysisWorkshop/blob/main/vignettes/workflow_spatial_cosmx.Rmd" class="external-link"><code>vignettes/workflow_spatial_cosmx.Rmd</code></a></small>
      <div class="hidden name"><code>workflow_spatial_cosmx.Rmd</code></div>

    </div>

    
    
<p>
</p>
<p><strong>R version</strong>: R version 4.5.1 (2025-06-13) <br><strong>Bioconductor version</strong>: 3.21 <br><strong>Package
version</strong>: 1.0.0</p>

<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Spatial molecular technologies have revolutionised the study of
biological systems. We can now study cellular and sub-cellular
structures in complex systems. However, these technologies generate
complex datasets that require new computational approaches and ideas to
analyse. Many of these pipelines are still being developed therefore
what is state-of-the-art now will not remain so for too long. However,
the biological aim of analysis will mostly remain the same with more
being uncovered as computational methods advance. This workshop intends
to demonstrate the types of biological questions that can be studied
using spatial transcriptomics datasets. Our intention is NOT to use the
state-of-the-art methods, but to show you how data can be analysed to
study spatial hypotheses. As users get familiar with these ideologies,
they can swap out older tools with newer, more powerful tools. As
instructors, we hope to update this workshop as more and better tools
become available. We look forward to receiving community contributions
as well.</p>
</div>
<div class="section level2">
<h2 id="pre-processing-steps-already-performed">Pre-processing steps already performed<a class="anchor" aria-label="anchor" href="#pre-processing-steps-already-performed"></a>
</h2>
<p>Data from the instrument comes in raw images that are often
summerised to transcript tables. These tables are then coupled with
immunofluorescence images for a few key proteins or marker such as DAPI
which stains the nucleus, Pan-Cytokeratin which stains the membrane of
cancer cells, a cocktail of membrane proteins, markers such as CD45
which stain immune cells, and a cocktail of cytoplasmic markers
(technology specific). These proteins are used to perform cell
segmentation as the measurements are obtained at the transcript level
and not the cell level (as is the case with single-cell RNA-seq data).
Platforms provide segmented data and users can opt to perform their own
segmentation if needed. If you wish you re-segment the data yourself,
you could try using methods such as BIDCell <span class="citation">(Fu
et al. 2024)</span> that use a single cell reference dataset to perform
segmentation using the marker proteins as well as transcriptomics
measurements. Do note that vendors are constantly improving segmentation
and recent improvements are accurate on difficult to segment cells such
as neurons and astrocytes.</p>
<p>We will use the already segmented cell-level count data. We will also
use a previously prepared SpatialExperiment object. If users want to
prepare their own object, all they need is the cell matrix file, the
feature table, the cell table and a table of spatial coordinates (if
this information is not already in the cell table). With these,
information on preparing your own SpatialExperiment object can be found
in this package’s <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/SpatialExperiment/inst/doc/SpatialExperiment.html" class="external-link">vignette</a>.
The design for this object is shown in the figure below.</p>
<div class="float">
<img src="https://raw.githubusercontent.com/drighelli/SpatialExperiment/refs/heads/devel/vignettes/SPE.png" alt="SpatialExperiment object schematic"><div class="figcaption">SpatialExperiment object schematic</div>
</div>
</div>
<div class="section level2">
<h2 id="loading-prepared-data">Loading prepared data<a class="anchor" aria-label="anchor" href="#loading-prepared-data"></a>
</h2>
<p>The data we will use in this workshop is from the 10x Xenium
platform. Since the objective of this workshop is to introduce you to
spatial transcriptomics analysis, we do not work with the full dataset,
which contains over 500,000 cells and instead opt to work with a smaller
subset of the sample. This region contains interesting features and is
therefore enough to showcase a complete spatial transcriptomics analysis
workflow. The data used in this workshop is a window with the x
coordinate in the interval (4800, 5800) and the y coordinate in the
interval (8500, 9500). This window is shown in the context of the full
tissue below.</p>
<div class="float">
<img src="https://raw.githubusercontent.com/bhuvad/XeniumSpatialAnalysisWorkshop/refs/heads/main/inst/extdata/full_tissue.png" alt="Full IDC"><div class="figcaption">Full IDC</div>
</div>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/drighelli/SpatialExperiment" class="external-link">SpatialExperiment</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">lung5</span><span class="op">)</span> <span class="co"># replicate 1</span></span>
<span><span class="co"># preview object</span></span>
<span><span class="va">lung5</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SpatialExperiment </span></span>
<span><span class="co">## dim: 980 13463 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): counts</span></span>
<span><span class="co">## rownames(980): VTN TGFBR2 ... ADGRG2 NegPrb8</span></span>
<span><span class="co">## rowData names(2): gene genetype</span></span>
<span><span class="co">## colnames(13463): Lung5_Rep3_16_1 Lung5_Rep3_16_10 ... Lung5_Rep3_22_998</span></span>
<span><span class="co">##   Lung5_Rep3_22_999</span></span>
<span><span class="co">## colData names(7): sample_id cell_id ... Level2 fov</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span>
<span><span class="co">## spatialCoords names(2) : x y</span></span>
<span><span class="co">## imgData names(1): sample_id</span></span></code></pre>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># retrieve column (cell) annotations</span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">lung5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 13463 rows and 7 columns</span></span>
<span><span class="co">##                      sample_id     cell_id              region  technology</span></span>
<span><span class="co">##                    &lt;character&gt; &lt;character&gt;         &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">## Lung5_Rep3_16_1     Lung5_Rep3        16_1              Tumour       CosMx</span></span>
<span><span class="co">## Lung5_Rep3_16_10    Lung5_Rep3       16_10              Tumour       CosMx</span></span>
<span><span class="co">## Lung5_Rep3_16_100   Lung5_Rep3      16_100 Abnormal epithelium       CosMx</span></span>
<span><span class="co">## Lung5_Rep3_16_1000  Lung5_Rep3     16_1000 Abnormal epithelium       CosMx</span></span>
<span><span class="co">## Lung5_Rep3_16_1001  Lung5_Rep3     16_1001 Abnormal epithelium       CosMx</span></span>
<span><span class="co">## ...                        ...         ...                 ...         ...</span></span>
<span><span class="co">## Lung5_Rep3_22_995   Lung5_Rep3      22_995              Stroma       CosMx</span></span>
<span><span class="co">## Lung5_Rep3_22_996   Lung5_Rep3      22_996              Stroma       CosMx</span></span>
<span><span class="co">## Lung5_Rep3_22_997   Lung5_Rep3      22_997              Stroma       CosMx</span></span>
<span><span class="co">## Lung5_Rep3_22_998   Lung5_Rep3      22_998              Stroma       CosMx</span></span>
<span><span class="co">## Lung5_Rep3_22_999   Lung5_Rep3      22_999              Stroma       CosMx</span></span>
<span><span class="co">##                                    Level1              Level2         fov</span></span>
<span><span class="co">##                               &lt;character&gt;         &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">## Lung5_Rep3_16_1    Tumour area with str.. Epithelium - panCK+       fov16</span></span>
<span><span class="co">## Lung5_Rep3_16_10   Tumour area with str..              Stroma       fov16</span></span>
<span><span class="co">## Lung5_Rep3_16_100  Abnormal but not tum..              Stroma       fov16</span></span>
<span><span class="co">## Lung5_Rep3_16_1000 Abnormal but not tum..              Stroma       fov16</span></span>
<span><span class="co">## Lung5_Rep3_16_1001 Abnormal but not tum..              Stroma       fov16</span></span>
<span><span class="co">## ...                                   ...                 ...         ...</span></span>
<span><span class="co">## Lung5_Rep3_22_995                      NA              Stroma       fov22</span></span>
<span><span class="co">## Lung5_Rep3_22_996                      NA              Stroma       fov22</span></span>
<span><span class="co">## Lung5_Rep3_22_997                      NA              Stroma       fov22</span></span>
<span><span class="co">## Lung5_Rep3_22_998                      NA              Stroma       fov22</span></span>
<span><span class="co">## Lung5_Rep3_22_999                      NA              Stroma       fov22</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># retrieve row (gene/probe) annotations</span></span>
<span><span class="fu">rowData</span><span class="op">(</span><span class="va">lung5</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 980 rows and 2 columns</span></span>
<span><span class="co">##                gene    genetype</span></span>
<span><span class="co">##         &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">## VTN             VTN        Gene</span></span>
<span><span class="co">## TGFBR2       TGFBR2        Gene</span></span>
<span><span class="co">## JUNB           JUNB        Gene</span></span>
<span><span class="co">## FKBP11       FKBP11        Gene</span></span>
<span><span class="co">## REG1A         REG1A        Gene</span></span>
<span><span class="co">## ...             ...         ...</span></span>
<span><span class="co">## TSC22D1     TSC22D1        Gene</span></span>
<span><span class="co">## CD40           CD40        Gene</span></span>
<span><span class="co">## GDNF           GDNF        Gene</span></span>
<span><span class="co">## ADGRG2       ADGRG2        Gene</span></span>
<span><span class="co">## NegPrb8     NegPrb8      NegPrb</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># retrieve spatial coordinates</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">lung5</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                            x        y</span></span>
<span><span class="co">## Lung5_Rep3_16_1    -8679.934 12706.31</span></span>
<span><span class="co">## Lung5_Rep3_16_10   -8894.582 12709.94</span></span>
<span><span class="co">## Lung5_Rep3_16_100  -7189.803 12549.21</span></span>
<span><span class="co">## Lung5_Rep3_16_1000 -8028.512 11347.95</span></span>
<span><span class="co">## Lung5_Rep3_16_1001 -7470.037 11347.64</span></span>
<span><span class="co">## Lung5_Rep3_16_1002 -6631.154 11351.96</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># retrieve counts - top 5 cells and genes</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">lung5</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## 5 x 5 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">##        Lung5_Rep3_16_1 Lung5_Rep3_16_10 Lung5_Rep3_16_100 Lung5_Rep3_16_1000</span></span>
<span><span class="co">## VTN                  .                .                 2                  .</span></span>
<span><span class="co">## TGFBR2               .                1                 .                  .</span></span>
<span><span class="co">## JUNB                 .                1                 .                  1</span></span>
<span><span class="co">## FKBP11               .                .                 .                  .</span></span>
<span><span class="co">## REG1A                .                .                 .                  .</span></span>
<span><span class="co">##        Lung5_Rep3_16_1001</span></span>
<span><span class="co">## VTN                     .</span></span>
<span><span class="co">## TGFBR2                  .</span></span>
<span><span class="co">## JUNB                    .</span></span>
<span><span class="co">## FKBP11                  .</span></span>
<span><span class="co">## REG1A                   .</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bhuvad.github.io/SpaNorm" class="external-link">SpaNorm</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set default point size to 0.5 for this report</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/update_defaults.html" class="external-link">update_geom_defaults</a></span><span class="op">(</span>geom <span class="op">=</span> <span class="st">"point"</span>, new <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/plotSpatial.html" class="external-link">plotSpatial</a></span><span class="op">(</span><span class="va">lung5</span>, colour <span class="op">=</span> <span class="va">region</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co"># improve legend visibility</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">15</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_spatial_cosmx_files/figure-html/unnamed-chunk-2-1.png" width="700"></p>
<p>For users interested in analysing the whole dataset, it can be
obtained using the code below.</p>
<blockquote>
<p>Please note that the full dataset contains over 500,000 cells and
will not scale for the purpose of this workshop. If interested, run the
code below on your own laptop.</p>
</blockquote>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## DO NOT RUN during the workshop as the data are large and could crash your session</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"ExperimentHub"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"ExperimentHub"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st">"SubcellularSpatialData"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"SubcellularSpatialData"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># load the metadata from the ExperimentHub</span></span>
<span><span class="va">eh</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html" class="external-link">ExperimentHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># download the transcript-level data</span></span>
<span><span class="va">tx</span> <span class="op">=</span> <span class="va">eh</span><span class="op">[[</span><span class="st">"EH8232"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co"># filter the Lung5_Rep1 sample</span></span>
<span><span class="va">tx</span> <span class="op">=</span> <span class="va">tx</span><span class="op">[</span><span class="va">tx</span><span class="op">$</span><span class="va">sample_id</span> <span class="op">==</span> <span class="st">"Lung5_Rep1"</span>, <span class="op">]</span></span>
<span><span class="co"># summarise measurements for each cell</span></span>
<span><span class="va">lung5</span> <span class="op">=</span> <span class="fu">tx2spe</span><span class="op">(</span><span class="va">tx</span>, bin <span class="op">=</span> <span class="st">"cell"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quality-control">Quality control<a class="anchor" aria-label="anchor" href="#quality-control"></a>
</h2>
<p>Most spatial transcriptomics assays include negative control probes
to aid in quality control. These probes do not bind to the transcripts
of interest (signal), so represent noise in the assay. There are three
different types of negative controls in the 10x Xenium assay. As these
vary across technologies and serve different purposes, we treat them
equally for our analysis and assess them collectively. The breakdown of
the different types of negative control probes is given below.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="fu">rowData</span><span class="op">(</span><span class="va">lung5</span><span class="op">)</span><span class="op">$</span><span class="va">genetype</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">##   Gene NegPrb </span></span>
<span><span class="co">##    960     20</span></span></code></pre>
<p>We now use compute and visualise quality control metrics for each
cell. Some of the key metrics we assess are:</p>
<ol style="list-style-type: decimal">
<li>
<code>sum</code> or <code>total</code> - Total number of transcripts
and negative control probes per cell (more commonly called the library
size).</li>
<li>
<code>detected</code> - Number of transcripts and negative control
probes detected per cell (count &gt; 0).</li>
<li>
<code>subsets_neg_sum</code> - Count of negative control probes
(noise) per cell.</li>
<li>
<code>subsets_neg_detected</code> - Number of negative control
probes expressed per cell (count &gt; 0).</li>
<li>
<code>subsets_neg_percent</code> - <code>subsets_neg_sum</code> /
<code>sum</code>.</li>
</ol>
<p>Additionally, the vendor (10x) provides their own summary of quality
through the <code>qv</code> score. This is usually provided per detected
measurements but we compute the average per cell when deriving cellular
summaries. Likewise, we compute the number of transcripts allocated to
the nucleus (<code>overlaps_nucleus</code>).</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scuttle</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># identify the negative control probes</span></span>
<span><span class="va">is_neg</span> <span class="op">=</span> <span class="fu">rowData</span><span class="op">(</span><span class="va">lung5</span><span class="op">)</span><span class="op">$</span><span class="va">genetype</span> <span class="op">!=</span> <span class="st">"Gene"</span></span>
<span><span class="co"># compute QC metrics for each cell</span></span>
<span><span class="va">lung5</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQCMetrics</a></span><span class="op">(</span><span class="va">lung5</span>, subsets <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"neg"</span> <span class="op">=</span> <span class="va">is_neg</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<blockquote>
<p>Quality control is all about visualisation!</p>
</blockquote>
<p>One of the best ways to understand the data and identify problematic
cells or regions is to visualise the QC metrics. We need to visualise
the metrics in various contexts to better understand what cells appear
of poor qaulity. As such, we usually plot metrics against each other, we
visualise them in space, and we visualise them in the context of the
higher dimensional structure of the data. To aid the latter, we will
compute lower dimensional representations using the uniform manifold
approximation and projection (UMAP). As the UMAP embedding can be
computationally intensive, we usually perform PCA prior to UMAP. This
approach to quality control is identical to that used for single-cell
RNA-seq analysis.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># compute log counts - to fix the skew with counts</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">lung5</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/dge.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">lung5</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="co"># compute PCA</span></span>
<span><span class="va">lung5</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">lung5</span>, ncomponents <span class="op">=</span> <span class="fl">30</span><span class="op">)</span></span>
<span><span class="co"># compute UMAP</span></span>
<span><span class="va">lung5</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">lung5</span>, dimred <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span></code></pre></div>
<p>We plot the <code>sum</code> (library size) against the number of
genes detected, and colouring the points based on the proportion of
negative control probes (<code>subsets_neg_percent</code>). We see that
as more genes are detected, more measurements are obtained (library size
increases). There are cells where the library size is very low and the
number of genes detected is low as well. Considering the smaller panel
size of 960 target genes in this assay, we expect the number of genes
detected in some cell types to be small. However, if we detect very few
genes in a cell, we are less likely to obtain any useful information
from them. As such, poor quality cells can be defined as those with very
low library sizes and number of genes detected.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html" class="external-link">plotColData</a></span><span class="op">(</span><span class="va">lung5</span>, <span class="st">"detected"</span>, <span class="st">"sum"</span>, colour_by <span class="op">=</span> <span class="st">"region"</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">15</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span>, title <span class="op">=</span> <span class="st">"Region"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">50</span>, lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">75</span>, lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_spatial_cosmx_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>However, we need to identify appropriate thresholds to classify poor
quality cells. As the panel sizes and compositions for these emerging
technologies are variable, there are no standards as of yet for
filtering. For instance, immune cells will always have smaller library
sizes and potentially a smaller number of genes expressed in a targetted
panel as each cell type may be represented by a few marker genes.
Additionally, these cells are smaller in size therefore could be
filtered out using an area-based filter. Desipte the lack of standards,
the extreme cases where cells express fewer than 20 genes and have fewer
than 50 detections are clearly of poor quality. As such, we can start
with these thresholds and adjust based on other metrics and
visualisations.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/DavisLaboratory/standR" class="external-link">standR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># UMAP plot</span></span>
<span><span class="va">p1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotDR.html" class="external-link">plotDR</a></span><span class="op">(</span><span class="va">lung5</span>, dimred <span class="op">=</span> <span class="st">"UMAP"</span>, colour <span class="op">=</span> <span class="va">sum</span>, alpha <span class="op">=</span> <span class="fl">0.75</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"F"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"UMAP1"</span>, y <span class="op">=</span> <span class="st">"UMAP2"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="co"># spatial plot</span></span>
<span><span class="va">p2</span> <span class="op">=</span> <span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/plotSpatial.html" class="external-link">plotSpatial</a></span><span class="op">(</span><span class="va">lung5</span>, colour <span class="op">=</span> <span class="va">sum</span>, alpha <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"F"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Sum (library size)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_spatial_cosmx_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># UMAP plot</span></span>
<span><span class="va">p1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotDR.html" class="external-link">plotDR</a></span><span class="op">(</span><span class="va">lung5</span>, dimred <span class="op">=</span> <span class="st">"UMAP"</span>, colour <span class="op">=</span> <span class="va">detected</span>, alpha <span class="op">=</span> <span class="fl">0.75</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"F"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"UMAP1"</span>, y <span class="op">=</span> <span class="st">"UMAP2"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="co"># spatial plot</span></span>
<span><span class="va">p2</span> <span class="op">=</span> <span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/plotSpatial.html" class="external-link">plotSpatial</a></span><span class="op">(</span><span class="va">lung5</span>, colour <span class="op">=</span> <span class="va">detected</span>, alpha <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"F"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Detected (number of genes expressed)"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_spatial_cosmx_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># UMAP plot</span></span>
<span><span class="va">p1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotDR.html" class="external-link">plotDR</a></span><span class="op">(</span><span class="va">lung5</span>, dimred <span class="op">=</span> <span class="st">"UMAP"</span>, colour <span class="op">=</span> <span class="va">subsets_neg_percent</span>, alpha <span class="op">=</span> <span class="fl">0.75</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"F"</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">5</span><span class="op">)</span>, oob <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/oob.html" class="external-link">squish</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"UMAP1"</span>, y <span class="op">=</span> <span class="st">"UMAP2"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="co"># spatial plot</span></span>
<span><span class="va">p2</span> <span class="op">=</span> <span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/plotSpatial.html" class="external-link">plotSpatial</a></span><span class="op">(</span><span class="va">lung5</span>, colour <span class="op">=</span> <span class="va">subsets_neg_percent</span>, alpha <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"F"</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">5</span><span class="op">)</span>, oob <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/oob.html" class="external-link">squish</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Proportion of negative control probes"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_spatial_cosmx_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>Having visualised the quality metrics and their distrition across
space and the genomic dimensions, we can now begin to test threshold
values. We want to remove low quality cells but we have to be careful
not to introduce bias (e.g., by preferentially removing biological
entities such as regions or cells).</p>
<p>Try modifying the thresholds below to see the effect of over- or
under- filtering.</p>
<ul>
<li>What happens when you set the <code>sum</code> threshold to
300?</li>
<li>What about setting the <code>detected</code> threshold to 100?</li>
</ul>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">thresh_sum</span> <span class="op">=</span> <span class="fl">75</span></span>
<span><span class="va">thresh_detected</span> <span class="op">=</span> <span class="fl">50</span></span>
<span><span class="va">thresh_neg</span> <span class="op">=</span> <span class="fl">5</span></span>
<span></span>
<span><span class="co"># classify cells to keep</span></span>
<span><span class="va">lung5</span><span class="op">$</span><span class="va">keep</span> <span class="op">=</span> <span class="va">lung5</span><span class="op">$</span><span class="va">sum</span> <span class="op">&gt;</span> <span class="va">thresh_sum</span> <span class="op">&amp;</span> <span class="va">lung5</span><span class="op">$</span><span class="va">detected</span> <span class="op">&gt;</span> <span class="va">thresh_detected</span> <span class="op">&amp;</span> <span class="va">lung5</span><span class="op">$</span><span class="va">subsets_neg_percent</span> <span class="op">&lt;</span> <span class="va">thresh_neg</span></span>
<span></span>
<span><span class="co"># UMAP plot</span></span>
<span><span class="va">p1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotDR.html" class="external-link">plotDR</a></span><span class="op">(</span><span class="va">lung5</span>, dimred <span class="op">=</span> <span class="st">"UMAP"</span>, colour <span class="op">=</span> <span class="va">keep</span>, alpha <span class="op">=</span> <span class="fl">0.75</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"UMAP1"</span>, y <span class="op">=</span> <span class="st">"UMAP2"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">15</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># spatial plot</span></span>
<span><span class="va">p2</span> <span class="op">=</span> <span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/plotSpatial.html" class="external-link">plotSpatial</a></span><span class="op">(</span><span class="va">lung5</span>, colour <span class="op">=</span> <span class="va">keep</span>, alpha <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set1"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">15</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Cells retained after QC filtering"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_spatial_cosmx_files/figure-html/unnamed-chunk-11-1.png" width="700"></p>
<p>In practice, after poor quality cells are filtered out, we tend to
see clearer separation of clusters in the UMAP (or equivalent
visualisations). We also notice that poor quality cells often seem like
intermmediate states between clusters in these plots. Additionally, they
tend to be randomly distributed in space. If they are concentrated in
some sub-regions (as above - along the field of view boundaries), it is
usually due to technical reasons. In the above plot, we poor quality
primarily occurs along the field of view boundaries. It is unlikely that
biology would be present in such a regular shape therefore we can use
that region as a guide in selecting thresholds. Having selected the
thresholds, we now apply them and remove the negative control probes
from donwstream analysis. This phenomenon of lower library size and
lower number of genes detected has been observed in the past and is due
to cells being cut off from the field of view.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># apply the filtering</span></span>
<span><span class="va">lung5</span> <span class="op">=</span> <span class="va">lung5</span><span class="op">[</span>, <span class="va">lung5</span><span class="op">$</span><span class="va">keep</span><span class="op">]</span></span>
<span><span class="co"># remove negative probes</span></span>
<span><span class="va">lung5</span> <span class="op">=</span> <span class="va">lung5</span><span class="op">[</span><span class="op">!</span><span class="va">is_neg</span>, <span class="op">]</span></span></code></pre></div>
<p>Newer quality control tools are being developed and are currently
under peer-review. These include SpotSweeper <span class="citation">(Totty, Hicks, and Guo 2024)</span> and <a href="https://github.com/drighelli/SpaceTrooper" class="external-link">SpaceTrooper</a>. As
these software are not deposited to more permanent repositories, we do
not use them in this workshop. We advise advanced users to use these
tools for their QC as they help come up with additional QC summaries for
each cell or region in space. SpaceTrooper for instance, computes a flag
score that penalises cells that are too large or small, and whose aspect
ratio is large. While these filters may make sense in most systems, they
could penalise stomal or neuronal cells that are elongated and whose
aspect ratio will be very large. Therefore always remember that QC is
relative the biological system being studied.</p>
<p>As these technologies are still emerging, all quality control metrics
should be evaluated on a case-by-case basis depending on the platform,
panel set, and tissue. Knowing the biology to expect or what is
definitely NOT biology helps in making these decisions. As any
experienced bioinformatician will tell you, you will vary these
thresholds a few times throughout your analysis because you learn more
about the data, biology, and platform as you analyse the data. This
cyclical approach can introduce biases so always rely on multiple
avenues of evidence before changing parameters.</p>
</div>
<div class="section level2">
<h2 id="normalisation">Normalisation<a class="anchor" aria-label="anchor" href="#normalisation"></a>
</h2>
<p>Above we plot the <code>sum</code>, which is commonly known as the
library size, for the purpose of quality control. Assessing the library
size is helpful in identifying poor quality cells. Beyond QC, variation
in library size can affect donwstream analysis as it represents
measurement bias. A gene may appear to be highly expressed in some cells
simply because more measurements were sampled from the cell, and not
because the gene is truly highly expressed. Whether it is
sequencing-based or imaging-based transcriptomics, we are always
sampling from the true pool of transcripts and are therefore likely to
have library size variation.</p>
<p>While library sizes are a good measure of technical variation in
conventional sequencing data (bulk and single-cell RNA-seq), they are
not the most appropriate in assessing technical effects in targeted
panels because the panel is biased towards measuring the biology of
interest. As such, we need a purely technical measurement to adjust our
data. Negative control probes offer this measure of technical biases in
the data as they are designed to not measure any biology (i.e., the are
synthetic probes designed to not hybridise with any transcripts in the
system).</p>
<p>Though it is ideal to use negative control probes for normalisation
of targeted spatial transcriptomics data, the data we are analysing is
relatively old and from a prototype instrument, therefore, the negative
control probe counts are too low to be used for modelling. Additionally,
the current Bruker CosMx instrument uses two kinds of negative control
probes, conventional negative control probes (such as in this dataset)
and false codes. The latter are best used for modelling technical
variation in CosMx datasets as their counts are generally higher. We
recommend using false code counts for normalising Bruker CosMx
datasets.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lung5</span><span class="op">$</span><span class="va">logLS</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="va">lung5</span><span class="op">$</span><span class="va">sum</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log1p</a></span><span class="op">(</span><span class="va">lung5</span><span class="op">$</span><span class="va">sum</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/plotSpatial.html" class="external-link">plotSpatial</a></span><span class="op">(</span><span class="va">lung5</span>, colour <span class="op">=</span> <span class="va">logLS</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_colour_distiller</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"PRGn"</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.5</span>, <span class="fl">0.5</span><span class="op">)</span>, oob <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/oob.html" class="external-link">squish</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_spatial_cosmx_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>To remove these technical biases, we perform normalisation. As
measurements are independent in bulk and single-cell RNA-seq (reactions
happen independently for each sample/cell), each cell or sample can be
normalised independently by computing sample/cell-specific adjustment
factors. Measurements are locally dependent in spatial transcriptomics
datasets therefore technical effects (including conventional library
size effects) confound biology <span class="citation">(Bhuva et al.
2024)</span>. The plot above show this clearly as the technical effect
demarcate tissue structures. These can be due to various causes such
as:</p>
<ol style="list-style-type: decimal">
<li>The tissue structure in one region is more rigid, thereby making it
difficult for reagents to permeate.</li>
<li>Reagents distributed unevenly due to technical issues leading to a
sub-region within a tissue region being under sampled.</li>
</ol>
<p>In both cases, each cell cannot be normalised independently as nearby
cells are likely to experience similar technical effects. To tackle
this, our team has developed the SpaNorm normalisation method that uses
spatial dependence to decouple technical and biological effects <span class="citation">(Salim et al. 2024)</span>. SpaNorm uses a spatially
constrained regularised generalised linear model to model library size
effects and subsequently adjust the data.</p>
<div class="float">
<img src="https://raw.githubusercontent.com/bhuvad/SpaNorm/refs/heads/master/vignettes/SpaNormWorkflow.png" alt="SpaNorm schematic"><div class="figcaption">SpaNorm schematic</div>
</div>
<p>The key parameters for SpaNorm normalisation that need to be adjusted
are:</p>
<ol style="list-style-type: decimal">
<li>
<code>sample.p</code> - This is the proportion of the data that is
sub-sampled to fit the regularised GLM. As spatial transcriptomics
datasets are very large, the model can be approximated using by
sampling. In practice, we see that ~50,000 cells in a dataset with
500,000 cells works well enough under the assumption that all possible
cellular states are sampled evenly across the tissue. This is likely to
be true in most cases. Below we use 25% of the cells (which is the
default as well) to fit the model (you can tweak this to 1 to use all
cells).</li>
<li>
<code>df.tps</code> - Spatial dependence is modelled using thin
plate spline functions. This parameter controls the degrees of freedom
(complexity) of the function. Complex functions will fit the data better
but could overfit therefore in practice, we see that 6-10 degrees of
freedom are generally good enough. If the degrees of freedom are
increased, regularisation should also be increased
(<code>lambda.a</code> parameter). We also allow separate specification
of the degrees of freedom for the spline functions for biological and
technical effects. Below we specify 8 d.f. for biology and 3 d.f. for
technical effects. This works on the assumption that biology will vary
more sharply across space than technical effects, which we expect to be
relatively smooth.</li>
</ol>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="co"># perform SpaNorm normalisation</span></span>
<span><span class="va">lung5</span> <span class="op">=</span> <span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/SpaNorm.html" class="external-link">SpaNorm</a></span><span class="op">(</span><span class="va">lung5</span>, df.tps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">8</span>, <span class="fl">3</span><span class="op">)</span>, sample.p <span class="op">=</span> <span class="fl">0.25</span>, batch <span class="op">=</span> <span class="va">lung5</span><span class="op">$</span><span class="va">fov</span><span class="op">)</span></span></code></pre></div>
<p>Above, the model is fit and the default adjustment,
<code>logPAC</code>, is computed. Alternative adjustments such as the
mean biology (<code>meanbio</code>), pearson residuals similar to
sctransform used in Seurat (<code>pearson</code>), and median biology
(<code>median</code>) can be computed as well (this will be faster as
the fit does not need to be repeated, however, the same parameters need
to be passed to the function).</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># compute mean biology</span></span>
<span><span class="va">lung5_pearson</span> <span class="op">=</span> <span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/SpaNorm.html" class="external-link">SpaNorm</a></span><span class="op">(</span><span class="va">lung5</span>, df.tps <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">8</span>, <span class="fl">3</span><span class="op">)</span>, batch <span class="op">=</span> <span class="va">lung5</span><span class="op">$</span><span class="va">fov</span>, adj.method <span class="op">=</span> <span class="st">"pearson"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p1</span> <span class="op">=</span> <span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/plotSpatial.html" class="external-link">plotSpatial</a></span><span class="op">(</span><span class="va">lung5</span>, what <span class="op">=</span> <span class="st">"expression"</span>, assay <span class="op">=</span> <span class="st">"logcounts"</span>, colour <span class="op">=</span> <span class="va">EPCAM</span>, alpha <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"F"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Normalised expression"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">=</span> <span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/plotSpatial.html" class="external-link">plotSpatial</a></span><span class="op">(</span><span class="va">lung5_pearson</span>, what <span class="op">=</span> <span class="st">"expression"</span>, assay <span class="op">=</span> <span class="st">"logcounts"</span>, colour <span class="op">=</span> <span class="va">EPCAM</span>, alpha <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"F"</span>, limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="fl">3</span><span class="op">)</span>, oob <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/oob.html" class="external-link">squish</a></span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Mean from SpaNorm model"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="workflow_spatial_cosmx_files/figure-html/unnamed-chunk-15-1.png" width="700"></p>
<p>Since the SpaNorm model learns the functions for biological and
technical effects for each gene, we can also investigate these to ensure
that the correct biology is learnt and modelled. The
<code>plotCovariate</code> functions can be used to study these. It is
worth remembering that while the biological function is independent of
covariates (like an intercept), the technical function is a function of
the library size. To enable some interpretation, we show how the
function evaluates at the mean library size below.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">=</span> <span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/plotCovariate.html" class="external-link">plotCovariate</a></span><span class="op">(</span><span class="va">lung5</span>, covariate <span class="op">=</span> <span class="st">"biology"</span>, colour <span class="op">=</span> <span class="va">EPCAM</span>, alpha <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"F"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Biological variation"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">=</span> <span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/plotCovariate.html" class="external-link">plotCovariate</a></span><span class="op">(</span><span class="va">lung5</span>, covariate <span class="op">=</span> <span class="st">"ls"</span>, colour <span class="op">=</span> <span class="va">EPCAM</span>, alpha <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"G"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Technical variation"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="va">p3</span> <span class="op">=</span> <span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/plotCovariate.html" class="external-link">plotCovariate</a></span><span class="op">(</span><span class="va">lung5</span>, covariate <span class="op">=</span> <span class="st">"batch"</span>, colour <span class="op">=</span> <span class="va">EPCAM</span>, alpha <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"A"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Batch variation"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span> <span class="op">+</span> <span class="va">p3</span></span></code></pre></div>
<p><img src="workflow_spatial_cosmx_files/figure-html/unnamed-chunk-16-1.png" width="768"></p>
<p><a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=DHRS" class="external-link"><em>EPCAM</em></a>
is an epithelial cell marker gene that will also be expressed in lung
cancer cells that are of epithelial origin. The SpaNorm model correctly
models the regions where the cancer cells exist (mean) and is able to
normalise technical effects without affecting the biology. The technical
function looks similar to the function for biology and this is expected
as technical effects due to library size biases confound biology,
however, SpaNorm is able to perform the adjustment while preserving most
of the biology.</p>
<p>Library size normalisation is important to ensure technical variation
is removed from gene expression. In the most extreme cases, a gene that
is highly expressed in a region of interest may seem to be lowly
expressed because of a drop in library sizes in said region (see
manuscript). We have seen that negative probes across most technologies
are also capturing library size effects very well and are experimenting
with using these to model technical variation and normalise data. In
some technologies/datasets, these are very lowly expressed, not because
of better data quality but by design as well, and therefore would NOT
faithfully represent technical variation. As with every other step, this
is still an active area of research.</p>
</div>
<div class="section level2">
<h2 id="identifying-spatially-variable-genes-svgs">Identifying spatially variable genes (SVGs)<a class="anchor" aria-label="anchor" href="#identifying-spatially-variable-genes-svgs"></a>
</h2>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">lung5</span> <span class="op">=</span> <span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/SpaNormSVG.html" class="external-link">SpaNormSVG</a></span><span class="op">(</span><span class="va">lung5</span><span class="op">)</span></span>
<span><span class="va">topsvgs</span> <span class="op">=</span> <span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/topSVGs.html" class="external-link">topSVGs</a></span><span class="op">(</span><span class="va">lung5</span>, n <span class="op">=</span> <span class="fl">9</span>, fdr <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">topsvgs</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">gene</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/plotSpatial.html" class="external-link">plotSpatial</a></span><span class="op">(</span></span>
<span>    <span class="va">lung5</span>,</span>
<span>    what <span class="op">=</span> <span class="st">"expression"</span>,</span>
<span>    assay <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>    colour <span class="op">=</span> <span class="op">!</span><span class="op">!</span><span class="fu"><a href="https://rlang.r-lib.org/reference/sym.html" class="external-link">sym</a></span><span class="op">(</span><span class="va">gene</span><span class="op">)</span>,</span>
<span>    size <span class="op">=</span> <span class="fl">0.2</span></span>
<span>  <span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"F"</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span>ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_spatial_cosmx_files/figure-html/unnamed-chunk-18-1.png" width="960"></p>
</div>
<div class="section level2">
<h2 id="glm-pca">GLM-PCA<a class="anchor" aria-label="anchor" href="#glm-pca"></a>
</h2>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lung5</span> <span class="op">=</span> <span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/SpaNormPCA.html" class="external-link">SpaNormPCA</a></span><span class="op">(</span><span class="va">lung5</span>, svg.fdr <span class="op">=</span> <span class="fl">0.1</span>, ncomponents <span class="op">=</span> <span class="fl">20</span>, residuals <span class="op">=</span> <span class="st">"pearson"</span><span class="op">)</span></span>
<span><span class="va">lung5</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">lung5</span>, min_dist <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># UMAP plot</span></span>
<span><span class="va">p1</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotDR.html" class="external-link">plotDR</a></span><span class="op">(</span><span class="va">lung5</span>, dimred <span class="op">=</span> <span class="st">"UMAP"</span>, colour <span class="op">=</span> <span class="va">region</span>, alpha <span class="op">=</span> <span class="fl">0.75</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">15</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"UMAP1"</span>, y <span class="op">=</span> <span class="st">"UMAP2"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/standR/man/plotDR.html" class="external-link">plotDR</a></span><span class="op">(</span><span class="va">lung5</span>, dimred <span class="op">=</span> <span class="st">"UMAP"</span>, colour <span class="op">=</span> <span class="va">fov</span>, alpha <span class="op">=</span> <span class="fl">0.75</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Set2"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">15</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"UMAP1"</span>, y <span class="op">=</span> <span class="st">"UMAP2"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span></span></code></pre></div>
<p><img src="workflow_spatial_cosmx_files/figure-html/unnamed-chunk-19-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="cell-typing">Cell typing<a class="anchor" aria-label="anchor" href="#cell-typing"></a>
</h2>
<p>Cell typing can be performed using various approaches. Most of these
were developed for the analysis of single-cell RNA-seq datasets and port
across well to spatial transcriptomics datasets. The two common
strategies are reference-based and reference-free annotations. In the
former, annotations are mapped across from a previously annotated
dataset. For a thorough discussion on this topic, review the <a href="https://bioconductor.org/books/release/OSCA/" class="external-link">Orchestrating
Single-Cell Analysis (OSCA) with Bioconductor</a> book <span class="citation">(Amezquita et al. 2020)</span>. Here, we use the
SingleR package <span class="citation">(Aran et al. 2019)</span> along
with a reference non-small cell lung cancer (NSCLC) dataset <span class="citation">(<strong>salcher2022high?</strong>)</span>.</p>
<p>SingleR is a simple yet powerful algorithm that only requires a
single prototype of what the cell type’s expression profile should look
like. As such, counts are aggregated (pseudobulked) across cells to form
a reference dataset where a single expression vector is created for each
cell type. Aggregation is performed using the
<code><a href="https://rdrr.io/pkg/scuttle/man/aggregateAcrossCells.html" class="external-link">scater::aggregateAcrossCells</a></code> function. As the original
dataset contains &gt;96,000 cells, we only provide the aggregated counts
for this workshop. Aggregated counts tend to be more robust to technical
variation and are therefore more suitable for cell typing.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/MarioniLab/scran/" class="external-link">scran</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load reference dataset - Salcher et al., Cancer Cell, 2022</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">ref_luca</span><span class="op">)</span></span>
<span><span class="co"># normalise the reference dataset</span></span>
<span><span class="va">cl</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/quickCluster.html" class="external-link">quickCluster</a></span><span class="op">(</span><span class="va">ref_luca</span><span class="op">)</span></span>
<span><span class="va">ref_luca</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html" class="external-link">computeSumFactors</a></span><span class="op">(</span><span class="va">ref_luca</span>, cluster <span class="op">=</span> <span class="va">cl</span><span class="op">)</span></span>
<span><span class="va">ref_luca</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">ref_luca</span><span class="op">)</span></span>
<span><span class="co"># cell types in the reference dataset</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">ref_luca</span><span class="op">$</span><span class="va">cell_type</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "Alveolar cell type 1"       "Alveolar cell type 2"      </span></span>
<span><span class="co">##  [3] "B cell"                     "B cell dividing"           </span></span>
<span><span class="co">##  [5] "Ciliated"                   "Club"                      </span></span>
<span><span class="co">##  [7] "DC mature"                  "Endothelial cell arterial" </span></span>
<span><span class="co">##  [9] "Endothelial cell capillary" "Endothelial cell lymphatic"</span></span>
<span><span class="co">## [11] "Endothelial cell venous"    "Fibroblast adventitial"    </span></span>
<span><span class="co">## [13] "Fibroblast alveolar"        "Fibroblast peribronchial"  </span></span>
<span><span class="co">## [15] "Macrophage"                 "Macrophage alveolar"       </span></span>
<span><span class="co">## [17] "Mast cell"                  "Mesothelial"               </span></span>
<span><span class="co">## [19] "Monocyte classical"         "Monocyte non-classical"    </span></span>
<span><span class="co">## [21] "NK cell"                    "Neutrophils"               </span></span>
<span><span class="co">## [23] "Pericyte"                   "Plasma cell"               </span></span>
<span><span class="co">## [25] "Plasma cell dividing"       "ROS1+ healthy epithelial"  </span></span>
<span><span class="co">## [27] "Smooth muscle cell"         "T cell CD4"                </span></span>
<span><span class="co">## [29] "T cell CD8"                 "T cell dividing"           </span></span>
<span><span class="co">## [31] "T cell regulatory"          "Tumor cells"               </span></span>
<span><span class="co">## [33] "cDC1"                       "cDC2"                      </span></span>
<span><span class="co">## [35] "myeloid dividing"           "pDC"                       </span></span>
<span><span class="co">## [37] "stromal dividing"           "transitional club/AT2"</span></span></code></pre>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/SingleR-inc/SingleR" class="external-link">SingleR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># predict cell types using SingleR</span></span>
<span><span class="va">prediction</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleR/man/SingleR.html" class="external-link">SingleR</a></span><span class="op">(</span>test <span class="op">=</span> <span class="va">lung5</span>, ref <span class="op">=</span> <span class="va">ref_luca</span>, labels <span class="op">=</span> <span class="va">ref_luca</span><span class="op">$</span><span class="va">cell_type</span><span class="op">)</span></span>
<span><span class="co"># add prediction to the SpatialExperiment object</span></span>
<span><span class="va">lung5</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">=</span> <span class="va">prediction</span><span class="op">$</span><span class="va">labels</span></span></code></pre></div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/plotSpatial.html" class="external-link">plotSpatial</a></span><span class="op">(</span><span class="va">lung5</span>, colour <span class="op">=</span> <span class="va">cell_type</span>, alpha <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Predicted cell type"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">15</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_spatial_cosmx_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="spatial-domain-identification">Spatial domain identification<a class="anchor" aria-label="anchor" href="#spatial-domain-identification"></a>
</h2>
<p>Regions were annotated by an expert pathologist in our dataset.
However, this is not always feasible, either due to access to a
pathologist, or because the features are not visible in histology images
alone. In such instances, a computational approach is required to
identify spatial domains or cellular niches. These domains are
generally, but not always, defined based on the homogeneity of biology.
As this is still a developing area, there are many algorithms available,
however, they generally fall within two broad categories of methods:</p>
<ol style="list-style-type: decimal">
<li>Cell type-based - These methods use the cell type information to
identify regions where the cell type composition is homogeneous.</li>
<li>Expression-based - These methods use the full expression profile of
cells to identify regions of the tissue with homogeneous gene expression
profiles.</li>
</ol>
<div class="section level3">
<h3 id="approach-1---cell-type-based-domain-identification">Approach 1 - Cell type-based domain identification<a class="anchor" aria-label="anchor" href="#approach-1---cell-type-based-domain-identification"></a>
</h3>
<p>These methods use the cell types annotated in the previous type to
determine the homogeneity of cell types within a local region of space
around each cell. The <a href="https://bioconductor.org/packages/release/bioc/html/hoodscanR.html" class="external-link">hoodscanR</a>
method is one such method that computes the probability of each cell
type being in a local neighbourhood of each cell in the dataset <span class="citation">(Liu et al. 2024)</span>. The neighbourhood of each
cell is determined based on the <code>k</code> parameter which specifies
how many of the nearest neighbouring cells should be used to estimate
the cell type probabilities.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/DavisLaboratory/hoodscanR" class="external-link">hoodscanR</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># find neighbouring cells using an approximate nearest neighbour search</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">nb_cells</span> <span class="op">=</span> <span class="fu"><a href="https://davislaboratory.github.io/hoodscanR/reference/findNearCells.html" class="external-link">findNearCells</a></span><span class="op">(</span><span class="va">lung5</span>, k <span class="op">=</span> <span class="fl">100</span>, anno_col <span class="op">=</span> <span class="st">"cell_type"</span><span class="op">)</span></span>
<span><span class="va">nb_cells</span><span class="op">$</span><span class="va">cells</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##                              nearest_cell_1           nearest_cell_2</span></span>
<span><span class="co">## Lung5_Rep3_16_1    Fibroblast peribronchial Fibroblast peribronchial</span></span>
<span><span class="co">## Lung5_Rep3_16_10   Fibroblast peribronchial  Endothelial cell venous</span></span>
<span><span class="co">## Lung5_Rep3_16_100   Endothelial cell venous      Fibroblast alveolar</span></span>
<span><span class="co">## Lung5_Rep3_16_1000        T cell regulatory Fibroblast peribronchial</span></span>
<span><span class="co">## Lung5_Rep3_16_1001                   B cell               Macrophage</span></span>
<span><span class="co">##                              nearest_cell_3      nearest_cell_4</span></span>
<span><span class="co">## Lung5_Rep3_16_1         Fibroblast alveolar Fibroblast alveolar</span></span>
<span><span class="co">## Lung5_Rep3_16_10        Fibroblast alveolar Fibroblast alveolar</span></span>
<span><span class="co">## Lung5_Rep3_16_100    Monocyte non-classical   T cell regulatory</span></span>
<span><span class="co">## Lung5_Rep3_16_1000 Fibroblast peribronchial         Plasma cell</span></span>
<span><span class="co">## Lung5_Rep3_16_1001        T cell regulatory    myeloid dividing</span></span>
<span><span class="co">##                              nearest_cell_5</span></span>
<span><span class="co">## Lung5_Rep3_16_1    Fibroblast peribronchial</span></span>
<span><span class="co">## Lung5_Rep3_16_10        Fibroblast alveolar</span></span>
<span><span class="co">## Lung5_Rep3_16_100                 Mast cell</span></span>
<span><span class="co">## Lung5_Rep3_16_1000              Plasma cell</span></span>
<span><span class="co">## Lung5_Rep3_16_1001  Endothelial cell venous</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nb_cells</span><span class="op">$</span><span class="va">distance</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##                    nearest_cell_1 nearest_cell_2 nearest_cell_3 nearest_cell_4</span></span>
<span><span class="co">## Lung5_Rep3_16_1          59.66036       71.23732       93.28319      105.50952</span></span>
<span><span class="co">## Lung5_Rep3_16_10         79.42413       80.90802       87.58466      103.09469</span></span>
<span><span class="co">## Lung5_Rep3_16_100        42.83032       47.33316       66.48095       76.16124</span></span>
<span><span class="co">## Lung5_Rep3_16_1000       58.31700       66.59987       66.92916       69.88571</span></span>
<span><span class="co">## Lung5_Rep3_16_1001       65.97415       69.78375       85.52106       89.67675</span></span>
<span><span class="co">##                    nearest_cell_5</span></span>
<span><span class="co">## Lung5_Rep3_16_1         116.67380</span></span>
<span><span class="co">## Lung5_Rep3_16_10        110.36708</span></span>
<span><span class="co">## Lung5_Rep3_16_100        82.58831</span></span>
<span><span class="co">## Lung5_Rep3_16_1000      100.66268</span></span>
<span><span class="co">## Lung5_Rep3_16_1001      103.71212</span></span></code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># estimate proportion of each cell type in the neighbourhood of each cell</span></span>
<span><span class="va">probmat</span> <span class="op">=</span> <span class="fu"><a href="https://davislaboratory.github.io/hoodscanR/reference/scanHoods.html" class="external-link">scanHoods</a></span><span class="op">(</span><span class="va">nb_cells</span><span class="op">$</span><span class="va">distance</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">probmat</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 11721   100</span></span></code></pre>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># compute probability of each each cell type in the neighbour hood of each cell</span></span>
<span><span class="va">neighbourhoods</span> <span class="op">=</span> <span class="fu"><a href="https://davislaboratory.github.io/hoodscanR/reference/mergeByGroup.html" class="external-link">mergeByGroup</a></span><span class="op">(</span><span class="va">probmat</span>, <span class="va">nb_cells</span><span class="op">$</span><span class="va">cells</span><span class="op">)</span></span>
<span><span class="va">neighbourhoods</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##                    Alveolar cell type 1 Alveolar cell type 2      B cell</span></span>
<span><span class="co">## Lung5_Rep3_16_1            1.360625e-02         1.142060e-02 0.000000000</span></span>
<span><span class="co">## Lung5_Rep3_16_10           1.522646e-03         3.848242e-02 0.000000000</span></span>
<span><span class="co">## Lung5_Rep3_16_100          4.929518e-09         8.189433e-05 0.003586059</span></span>
<span><span class="co">## Lung5_Rep3_16_1000         0.000000e+00         0.000000e+00 0.009200223</span></span>
<span><span class="co">## Lung5_Rep3_16_1001         0.000000e+00         1.066744e-03 0.110972648</span></span>
<span><span class="co">##                    B cell dividing         cDC1</span></span>
<span><span class="co">## Lung5_Rep3_16_1        0.000000000 0.0283521102</span></span>
<span><span class="co">## Lung5_Rep3_16_10       0.000000000 0.0021066222</span></span>
<span><span class="co">## Lung5_Rep3_16_100      0.002428671 0.0002870623</span></span>
<span><span class="co">## Lung5_Rep3_16_1000     0.001067195 0.0001393729</span></span>
<span><span class="co">## Lung5_Rep3_16_1001     0.015444471 0.0349382750</span></span></code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add neighbourhood information to the SpatialExperiment object</span></span>
<span><span class="va">lung5</span> <span class="op">=</span> <span class="fu"><a href="https://davislaboratory.github.io/hoodscanR/reference/mergeHoodSpe.html" class="external-link">mergeHoodSpe</a></span><span class="op">(</span><span class="va">lung5</span>, <span class="va">neighbourhoods</span><span class="op">)</span></span></code></pre></div>
<p>Once these probabilies are assigned around each cell, a secondary
clustering can be performed using any desired method. The package
currently only supports k-means clustering on the neighbourhood matrix
but advanced users can cluster the matrix directly using their preferred
method.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># identify spatial domains based on neighbourhood composition (5 clusters)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">lung5</span> <span class="op">=</span> <span class="fu"><a href="https://davislaboratory.github.io/hoodscanR/reference/clustByHood.html" class="external-link">clustByHood</a></span><span class="op">(</span><span class="va">lung5</span>, pm_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">neighbourhoods</span><span class="op">)</span>, k <span class="op">=</span> <span class="fl">5</span>, val_name <span class="op">=</span> <span class="st">"hoodscanr_region"</span><span class="op">)</span></span>
<span><span class="co"># plot predictions</span></span>
<span><span class="va">p1</span> <span class="op">=</span> <span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/plotSpatial.html" class="external-link">plotSpatial</a></span><span class="op">(</span><span class="va">lung5</span>, colour <span class="op">=</span> <span class="va">hoodscanr_region</span>, alpha <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Paired"</span>, na.value <span class="op">=</span> <span class="st">"#333333"</span>, name <span class="op">=</span> <span class="st">"predicted"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Predicted pathology"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fl">3</span>, override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">15</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">=</span> <span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/plotSpatial.html" class="external-link">plotSpatial</a></span><span class="op">(</span><span class="va">lung5</span>, colour <span class="op">=</span> <span class="va">region</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Pathology annotations"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fl">3</span>, override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">15</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Spatial domain identification"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_spatial_cosmx_files/figure-html/unnamed-chunk-24-1.png" width="700"></p>
<p>The plot above shows that hoodscanR is able to differentiate the key
regions and is clearly able to separate the distinct DCIS lesions. It is
also able to identify the layer of cells separating the tumour lesions
from the stroma. However, it is unable to distinguish between the
invasive and DCIS regions. This is because our current annotation of
cell types does not discriminate between the different cellular states
of cells in these regions. This is one limitation of cell type based
methods - that is, they need the labels of all possible cell types and
cell states we are interested. However, the power of these methods is
that we can study the neighbourhood composition directly as this was
used to generate the domains.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://davislaboratory.github.io/hoodscanR/reference/plotProbDist.html" class="external-link">plotProbDist</a></span><span class="op">(</span><span class="va">lung5</span>, pm_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">neighbourhoods</span><span class="op">)</span>, by_cluster <span class="op">=</span> <span class="cn">TRUE</span>, plot_all <span class="op">=</span> <span class="cn">TRUE</span>, show_clusters <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">5</span><span class="op">)</span><span class="op">)</span>, val_name <span class="op">=</span> <span class="st">"hoodscanr_region"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_spatial_cosmx_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<p>we see that cluster 5 is entirely defined by the presence of
malignant cells while cluster 2 captures the tumour lining epithelial
cells, malignant cells, and fibroblasts. Cluster 1 is composed of the
immune cells, particularly T cells, B cells and macrophages, indicating
the formation of lymphoid aggregates. Lymphoid aggregates are
interesting structures in the context of solid cancers, and depending on
their expression profile, these can be beneficial for patients.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate metrics of heterogeneity - entropy</span></span>
<span><span class="va">lung5</span> <span class="op">=</span> <span class="fu"><a href="https://davislaboratory.github.io/hoodscanR/reference/calcMetrics.html" class="external-link">calcMetrics</a></span><span class="op">(</span><span class="va">lung5</span>, pm_cols <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">neighbourhoods</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># plot metric</span></span>
<span><span class="va">p1</span> <span class="op">=</span> <span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/plotSpatial.html" class="external-link">plotSpatial</a></span><span class="op">(</span><span class="va">lung5</span>, colour <span class="op">=</span> <span class="va">cell_type</span>, alpha <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Paired"</span>, na.value <span class="op">=</span> <span class="st">"#333333"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Predicted cell type"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">15</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">=</span> <span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/plotSpatial.html" class="external-link">plotSpatial</a></span><span class="op">(</span><span class="va">lung5</span>, colour <span class="op">=</span> <span class="va">entropy</span>, alpha <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_colour_viridis_c</a></span><span class="op">(</span>option <span class="op">=</span> <span class="st">"G"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Cell type heterogeneity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Cell type heterogeneity"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_spatial_cosmx_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
<p>Looking at the heterogeneity of cell types across space, we see that
cell composition is very homogeneous within the tumour regions while it
is highest in the lymphoid aggregates which are often characterised by
the presence of various different types of immune cells. We emphasise
again that the homogeneity of the tumour regions is determined based on
their cell type call alone and not on their expression profile. As such,
the tumours themselves could be heterogeneous if finer cell typing or
cell phenotyping were to be performed. Additionally, any errors made in
cell typing will propage to domain identification.</p>
<p>The approach that follows allows an unbiased identification of
domains using the gene expression data directly.</p>
</div>
<div class="section level3">
<h3 id="approach-2---expression-based-domain-identification">Approach 2 - Expression-based domain identification<a class="anchor" aria-label="anchor" href="#approach-2---expression-based-domain-identification"></a>
</h3>
<p>Expression-based methods can be further sub-categorised into three
categories:</p>
<ol style="list-style-type: decimal">
<li>Graph Neural Network (GNN) or Graph Convolutional Network (GCN)
based algorithms.</li>
<li>Probabilistic models based on markov random fields.</li>
<li>Neighbourhood diffusion followed by graph-based clustering.</li>
</ol>
<p>We note that this is not a comprehensive survey of the literature but
just a summary of the most popular methods in this space. The first two
approaches can be very powerful, however, are time consuming and will
not always scale with larger datasets. Diffusion followed by
graph-clustering is a relatively simple, yet powerful approach that
scales to large datasets. Banksy is one such algorithm that first shares
information within a local neighbourhood, and then performs a standard
single-cell RNA-seq inspired graph-based clustering to identify spatial
domains <span class="citation">(Singhal et al. 2024)</span>.</p>
<p>The following key parameters determine the final clustering
results:</p>
<ol style="list-style-type: decimal">
<li>
<code>lambda</code> - This controls the degree of information
sharing in the neighbourhood (range of [0, 1]) where larger values
result in higher diffusion and thereby smoother clusters.</li>
<li>
<code>resolution</code> - The resolution parameter for the
graph-based clustering using either the Louvain or Leiden algorithms.
Likewise, there are other parameters to control the graph-based
clustering. The OSCA book discusses choices for these parameters in
depth.</li>
<li>
<code>npcs</code> - The number of principal components to use to
condense the feature space. The OSCA book discusses choices for this in
depth.</li>
<li>
<code>k_geom</code> - The number of nearest neighbours used to share
information. The authors recommend values within [15, 30] and note that
variation is minimal within this range.</li>
</ol>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/prabhakarlab/Banksy" class="external-link">Banksy</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># parameters</span></span>
<span><span class="va">lambda</span> <span class="op">=</span> <span class="fl">0.2</span></span>
<span><span class="va">npcs</span> <span class="op">=</span> <span class="fl">30</span></span>
<span><span class="va">k_geom</span> <span class="op">=</span> <span class="fl">50</span></span>
<span><span class="va">res</span> <span class="op">=</span> <span class="fl">0.4</span></span>
<span></span>
<span><span class="co"># compute the features required by Banksy</span></span>
<span><span class="va">lung5</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Banksy/man/computeBanksy.html" class="external-link">computeBanksy</a></span><span class="op">(</span><span class="va">lung5</span>, assay_name <span class="op">=</span> <span class="st">"logcounts"</span>, k_geom <span class="op">=</span> <span class="va">k_geom</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># compute PCA on the features</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">lung5</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Banksy/man/runBanksyPCA.html" class="external-link">runBanksyPCA</a></span><span class="op">(</span><span class="va">lung5</span>, lambda <span class="op">=</span> <span class="va">lambda</span>, npcs <span class="op">=</span> <span class="va">npcs</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># perform graph-based clusering</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span></span>
<span><span class="va">lung5</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/Banksy/man/clusterBanksy.html" class="external-link">clusterBanksy</a></span><span class="op">(</span><span class="va">lung5</span>, lambda <span class="op">=</span> <span class="va">lambda</span>, npcs <span class="op">=</span> <span class="va">npcs</span>, resolution <span class="op">=</span> <span class="va">res</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># store to the column banksy_region</span></span>
<span><span class="va">lung5</span><span class="op">$</span><span class="va">banksy_region</span> <span class="op">=</span> <span class="va">lung5</span><span class="op">$</span><span class="va">`clust_M0_lam0.2_k50_res0.4`</span></span>
<span></span>
<span><span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/plotSpatial.html" class="external-link">plotSpatial</a></span><span class="op">(</span><span class="va">lung5</span>, colour <span class="op">=</span> <span class="va">banksy_region</span>, alpha <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Paired"</span>, na.value <span class="op">=</span> <span class="st">"#333333"</span>, name <span class="op">=</span> <span class="st">"predicted"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Mean from SpaNorm model"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">15</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_spatial_cosmx_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
<p>Try varying the parameters to see how the clustering changes. Do note
that the results are automatically stored in a new column in the
colData. Unfortunately, the function does not allow specification of the
new column name so the only approach is to look at the column names
using <code>colnames(colData(idc))</code> and find the name of the new
column (usually the last one) containing the clustering results. We then
store these to a column of our choosing to make it easier to reference
it.</p>
<p>Looking back at the original annotations, we see that clusters 3 and
5 represent the DCIS and Invasive tumours respectively. Similarly,
cluster 10 represents the blood vessel annotated by the pathologist.
Computational analysis reveals additional structures present in the
tissue that may differ only in the genomic measurements. Clustes 6 and 7
for instance seems to represent different structures in the stroma, some
of which may be visible in the histology image but are intestive to
label manually. We also find cluster 4 which represents a boundary of
cells surrounding invasive and DCIS lesions.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">=</span> <span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/plotSpatial.html" class="external-link">plotSpatial</a></span><span class="op">(</span><span class="va">lung5</span>, colour <span class="op">=</span> <span class="va">banksy_region</span>, alpha <span class="op">=</span> <span class="fl">0.75</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Paired"</span>, na.value <span class="op">=</span> <span class="st">"#333333"</span>, name <span class="op">=</span> <span class="st">"predicted"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Predicted pathology"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fl">3</span>, override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">15</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">=</span> <span class="fu"><a href="https://bhuvad.github.io/spaNorm/reference/plotSpatial.html" class="external-link">plotSpatial</a></span><span class="op">(</span><span class="va">lung5</span>, colour <span class="op">=</span> <span class="va">region</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"Pathology annotations"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>nrow <span class="op">=</span> <span class="fl">3</span>, override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">15</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p1</span> <span class="op">+</span> <span class="va">p2</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Spatial domain identification"</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_spatial_cosmx_files/figure-html/unnamed-chunk-28-1.png" width="700"></p>
<p>The choice of the better approach amongst the two demonstrated here
should be based on the question of interest. They both have their
strengths and limitations. BANKSY allows us to identify within tumour
differences because it assesses the expression profile directly,
however, it does not offer any additional interpretation of the domains
it identifies and is therefore normally coupled with differential
expression and differential compositional analysis. Errors in the domain
identification would propagate to these downstream tasks as well. On the
other hand, hoodscanR inherently provides interpretation of the domains
it identifies, but it is limited by the accuracy and resolution of cell
type annotation.</p>
</div>
</div>
<div class="section level2">
<h2 id="differential-expression">Differential expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h2>
<p>Cell type identification or clustering allows us to identify the
different cellular states in a biological system. The power of spatial
molecular datasets is that we can study how cellular states change in
response to different structures in the tissue. For instance, in our
tissue we see that there are two different types of cancer lesions,
ductal carcinoma in situ and invasive. Cancer cells in the latter lesion
tend to me aggresive and can spread more than those in the DCIS region.
We would be interested in knowning the biological processes that make
this happen.</p>
<p>Additionally, cancer cells interact with their environment including
cell types such as fibroblasts. We may be interested in identifying the
differences in fibroblasts that are in close proximity to cancer cells
as opposed to the distant ones. We can identify the genes that are
associated with these changes below.</p>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># create clusters to study how cell types vary across domains</span></span>
<span><span class="va">clusters</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">lung5</span><span class="op">$</span><span class="va">cell_type</span>, <span class="va">lung5</span><span class="op">$</span><span class="va">hoodscanr_region</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span></span>
<span><span class="va">clusters</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/table.html" class="external-link">table</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## clusters</span></span>
<span><span class="co">## Fibroblast peribronchial_5                     Club_2 </span></span>
<span><span class="co">##                        254                        345 </span></span>
<span><span class="co">##  Endothelial cell venous_3              Plasma cell_3 </span></span>
<span><span class="co">##                        382                        383 </span></span>
<span><span class="co">##               Macrophage_3 Fibroblast peribronchial_3 </span></span>
<span><span class="co">##                        405                        427 </span></span>
<span><span class="co">##                   B cell_4 Fibroblast peribronchial_1 </span></span>
<span><span class="co">##                        576                        584 </span></span>
<span><span class="co">##              Plasma cell_5              Tumor cells_2 </span></span>
<span><span class="co">##                        756                        772</span></span></code></pre>
<p>We see that the two largest clusters capture malignant cells in
spatial domains 2 and 5 which align with the invasive and DCIS regions
respectively based on the pathology annotations. We also see that though
most of the fibroblasts are dispersed in the stroma (region 6), those
that are closer to the cancer lesions are clustered in region 3. We
would be interested in studying the difference between these and
understanding whether they contain features that promote the tumour.</p>
<blockquote>
<p>The analysis below is only meant to be a demonstration of the
potential of spatial transcriptomics data. A proper statistical analysis
would require replicate tissue samples from across different patients.
In such a case, we would aggregate transcript counts across cell types
using the <code><a href="https://rdrr.io/pkg/scuttle/man/sumCountsAcrossCells.html" class="external-link">scater::sumCountsAcrossCells</a></code> to create
pseudo-bulk samples, and follow up with a differential expression
analysis using the <a href="https://doi.org/10.12688/f1000research.8987.2" class="external-link"><code>edgeR</code></a>
or <a href="https://doi.org/10.12688/f1000research.9005.3" class="external-link"><code>limma</code></a>
pipelines. Alternatively, a more advanced tool such as <a href="https://doi.org/10.1186/s13059-023-03159-6" class="external-link"><code>nicheDE</code></a>
could be used to perform analysis on a single tissue, however, it would
also be more robust with biological replicates. This tool is outside the
scope of this introductory workshop therefore is not covered.</p>
</blockquote>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># identify markers</span></span>
<span><span class="va">marker.info</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/scoreMarkers.html" class="external-link">scoreMarkers</a></span><span class="op">(</span><span class="va">lung5</span>, <span class="va">clusters</span>, pairings <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B cell_4"</span>, <span class="st">"T cell regulatory_4"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B cell_3"</span>, <span class="st">"T cell regulatory_3"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">marker.info</span></span></code></pre></div>
<pre><code><span><span class="co">## List of length 2</span></span>
<span><span class="co">## names(2): B cell_4 T cell regulatory_4</span></span></code></pre>
<p>Below we compare the malignant (cancer) cells from region 2
(invasive) against those from region 5 (DCIS). Investigating and
visialising the top genes shows that <a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=DHRS" class="external-link"><em>DHRS</em></a>
and <a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=MUC6" class="external-link"><em>MUC6</em></a>
are up-regulated in invasive cancer cells while <a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=GATA3" class="external-link"><em>GATA3</em></a>
and <a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=FLNB" class="external-link"><em>FLNB</em></a>
are moderately up-regulated in DCIS cancer cells.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># study markers of B cell_4</span></span>
<span><span class="va">chosen_bcell</span> <span class="op">=</span> <span class="va">marker.info</span><span class="op">[[</span><span class="st">"B cell_4"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">ordered_bcell</span> <span class="op">=</span> <span class="va">chosen_bcell</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">chosen_bcell</span><span class="op">$</span><span class="va">rank.AUC</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ordered_bcell</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 4 columns</span></span>
<span><span class="co">##        self.average other.average self.detected other.detected</span></span>
<span><span class="co">##           &lt;numeric&gt;     &lt;numeric&gt;     &lt;numeric&gt;      &lt;numeric&gt;</span></span>
<span><span class="co">## MALAT1     4.803664      4.459637      1.000000       0.994512</span></span>
<span><span class="co">## CD74       3.153887      2.312227      0.973958       0.848988</span></span>
<span><span class="co">## RPL34      1.797616      1.365601      0.814236       0.668511</span></span>
<span><span class="co">## MS4A1      1.416434      0.560667      0.631944       0.260054</span></span>
<span><span class="co">## IGHM       1.501242      0.702592      0.649306       0.298785</span></span>
<span><span class="co">## CD79A      0.777071      0.322778      0.477431       0.210886</span></span></code></pre>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotExpression.html" class="external-link">plotExpression</a></span><span class="op">(</span><span class="va">lung5</span>,</span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ordered_bcell</span><span class="op">)</span>, <span class="fl">9</span><span class="op">)</span>,</span>
<span>  x <span class="op">=</span> <span class="st">"hoodscanr_region"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"hoodscanr_region"</span>,</span>
<span>  ncol <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">4</span><span class="op">)</span>, lwd <span class="op">=</span> <span class="fl">0.1</span>, lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Paired"</span>, na.value <span class="op">=</span> <span class="st">"#333333"</span>, name <span class="op">=</span> <span class="st">"cluster"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">15</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_spatial_cosmx_files/figure-html/unnamed-chunk-31-1.png" width="768"></p>
<p>Next we compare the fibroblasts from region 6 (stroma) against those
from region 3 (tumour-adjacent). Investigating and visialising the top
genes shows that <a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=LUM" class="external-link"><em>LUM</em></a>
and <a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=POSTN" class="external-link"><em>POSTN</em></a>
are up-regulated in stromal fibroblasts while <a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=MMP11" class="external-link"><em>MMP11</em></a>
and <a href="https://www.genecards.org/cgi-bin/carddisp.pl?gene=AQP1" class="external-link"><em>AQP1</em></a>
are up-regulated in the tumour-adjacent fibroblasts.</p>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># study markers of T cell regulatory_4</span></span>
<span><span class="va">chosen_tcell</span> <span class="op">=</span> <span class="va">marker.info</span><span class="op">[[</span><span class="st">"T cell regulatory_4"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">ordered_tcell</span> <span class="op">=</span> <span class="va">chosen_tcell</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">chosen_tcell</span><span class="op">$</span><span class="va">rank.AUC</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ordered_tcell</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## DataFrame with 6 rows and 4 columns</span></span>
<span><span class="co">##       self.average other.average self.detected other.detected</span></span>
<span><span class="co">##          &lt;numeric&gt;     &lt;numeric&gt;     &lt;numeric&gt;      &lt;numeric&gt;</span></span>
<span><span class="co">## CD37      0.921737      0.921644      0.545455       0.478048</span></span>
<span><span class="co">## FYN       0.778307      0.329144      0.504132       0.213585</span></span>
<span><span class="co">## FYB1      0.712461      0.342734      0.454545       0.253396</span></span>
<span><span class="co">## IL6ST     0.721561      0.261453      0.429752       0.185830</span></span>
<span><span class="co">## IL32      1.029293      0.813102      0.611570       0.483851</span></span>
<span><span class="co">## TIGIT     0.557663      0.167929      0.396694       0.127260</span></span></code></pre>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotExpression.html" class="external-link">plotExpression</a></span><span class="op">(</span><span class="va">lung5</span>,</span>
<span>  features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ordered_tcell</span><span class="op">)</span>, <span class="fl">9</span><span class="op">)</span>,</span>
<span>  x <span class="op">=</span> <span class="st">"hoodscanr_region"</span>,</span>
<span>  colour_by <span class="op">=</span> <span class="st">"hoodscanr_region"</span>,</span>
<span>  ncol <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">4</span><span class="op">)</span>, lwd <span class="op">=</span> <span class="fl">0.1</span>, lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Paired"</span>, na.value <span class="op">=</span> <span class="st">"#333333"</span>, name <span class="op">=</span> <span class="st">"cluster"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">15</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="workflow_spatial_cosmx_files/figure-html/unnamed-chunk-32-1.png" width="768"></p>
</div>
<div class="section level2">
<h2 id="functional-analysis">Functional analysis<a class="anchor" aria-label="anchor" href="#functional-analysis"></a>
</h2>
<p>The example dataset used here was generated using the 10x Xenium
platform that measured 380 genes. More recent versions of this platform
and competing platforms such as the NanoString CosMx are now providing
around 5000 genes (soon whole transcriptome as well!). It is relatively
easy to scan through the strongest differentially expressed genes from a
list of at most 380 genes to understand the processes underpinning the
system. With larger panel sizes, we need computational tools to perform
such analyses, often referred to as gene-set enrichment analysis or
functional analysis. As this is an extensive topic, it is not covered in
depth in this workshop, however, we have previously developed a <a href="https://davislaboratory.github.io/GenesetAnalysisWorkflow/" class="external-link">workshop
on gene-set enrichment analysis</a> and refer the attendees to that if
they are interested. A non-programmatic interface of some of the tools
in our workshop can be accessed through <a href="https://visse.cloud/" class="external-link">vissE.cloud</a>.</p>
<blockquote>
<p>Optional: You can also explore the full dataset from this workshop on
vissE.cloud through the following <a href="https://visse.cloud/galleryexample/cosmx_NSCLCLung5Rep3_PCA" class="external-link">link</a>.
This exploratory analysis performs principal components analysis (PCA)
or non-negative matrix factorisation (NMF) on the dataset to identify
macroscopic biological programs. It then performs gene-set enrichment
analysis on the factor loadings to explain the biological processes.</p>
</blockquote>
</div>
<div class="section level2">
<h2 class="unnumbered" id="packages-used">Packages used<a class="anchor" aria-label="anchor" href="#packages-used"></a>
</h2>
<p>This workflow depends on various packages from version 3.21 of the
Bioconductor project, running on R version 4.5.1 (2025-06-13) or higher.
The complete list of the packages used for this workflow are shown
below:</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] Banksy_1.4.0                       hoodscanR_1.6.0                   </span></span>
<span><span class="co">##  [3] CosMxSpatialAnalysisWorkshop_1.0.0 singscore_1.28.0                  </span></span>
<span><span class="co">##  [5] SpaNorm_1.3.4                      SpatialExperiment_1.18.1          </span></span>
<span><span class="co">##  [7] SingleR_2.10.0                     standR_1.12.0                     </span></span>
<span><span class="co">##  [9] scater_1.36.0                      scran_1.36.0                      </span></span>
<span><span class="co">## [11] scuttle_1.18.0                     SingleCellExperiment_1.30.1       </span></span>
<span><span class="co">## [13] SummarizedExperiment_1.38.1        Biobase_2.68.0                    </span></span>
<span><span class="co">## [15] GenomicRanges_1.60.0               GenomeInfoDb_1.44.0               </span></span>
<span><span class="co">## [17] IRanges_2.42.0                     S4Vectors_0.46.0                  </span></span>
<span><span class="co">## [19] BiocGenerics_0.54.0                generics_0.1.4                    </span></span>
<span><span class="co">## [21] MatrixGenerics_1.20.0              matrixStats_1.5.0                 </span></span>
<span><span class="co">## [23] patchwork_1.3.1                    ggplot2_3.5.2                     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3        jsonlite_2.0.0           </span></span>
<span><span class="co">##   [3] magrittr_2.0.3            ggbeeswarm_0.7.2         </span></span>
<span><span class="co">##   [5] magick_2.8.7              farver_2.1.2             </span></span>
<span><span class="co">##   [7] rmarkdown_2.29            fs_1.6.6                 </span></span>
<span><span class="co">##   [9] ragg_1.4.0                vctrs_0.6.5              </span></span>
<span><span class="co">##  [11] memoise_2.0.1             DelayedMatrixStats_1.30.0</span></span>
<span><span class="co">##  [13] prettydoc_0.4.1           htmltools_0.5.8.1        </span></span>
<span><span class="co">##  [15] S4Arrays_1.8.1            BiocNeighbors_2.2.0      </span></span>
<span><span class="co">##  [17] SparseArray_1.8.0         sass_0.4.10              </span></span>
<span><span class="co">##  [19] bslib_0.9.0               htmlwidgets_1.6.4        </span></span>
<span><span class="co">##  [21] desc_1.4.3                plyr_1.8.9               </span></span>
<span><span class="co">##  [23] cachem_1.1.0              igraph_2.1.4             </span></span>
<span><span class="co">##  [25] lifecycle_1.0.4           pkgconfig_2.0.3          </span></span>
<span><span class="co">##  [27] rsvd_1.0.5                Matrix_1.7-3             </span></span>
<span><span class="co">##  [29] R6_2.6.1                  fastmap_1.2.0            </span></span>
<span><span class="co">##  [31] aricode_1.0.3             GenomeInfoDbData_1.2.14  </span></span>
<span><span class="co">##  [33] digest_0.6.37             AnnotationDbi_1.70.0     </span></span>
<span><span class="co">##  [35] dqrng_0.4.1               irlba_2.3.5.1            </span></span>
<span><span class="co">##  [37] textshaping_1.0.1         RSQLite_2.4.2            </span></span>
<span><span class="co">##  [39] beachmat_2.24.0           sccore_1.0.6             </span></span>
<span><span class="co">##  [41] labeling_0.4.3            httr_1.4.7               </span></span>
<span><span class="co">##  [43] abind_1.4-8               compiler_4.5.1           </span></span>
<span><span class="co">##  [45] bit64_4.6.0-1             withr_3.0.2              </span></span>
<span><span class="co">##  [47] BiocParallel_1.42.1       viridis_0.6.5            </span></span>
<span><span class="co">##  [49] DBI_1.2.3                 DelayedArray_0.34.1      </span></span>
<span><span class="co">##  [51] rjson_0.2.23              bluster_1.18.0           </span></span>
<span><span class="co">##  [53] tools_4.5.1               vipor_0.4.7              </span></span>
<span><span class="co">##  [55] beeswarm_0.4.0            glue_1.8.0               </span></span>
<span><span class="co">##  [57] dbscan_1.2.2              grid_4.5.1               </span></span>
<span><span class="co">##  [59] cluster_2.1.8.1           reshape2_1.4.4           </span></span>
<span><span class="co">##  [61] gtable_0.3.6              tidyr_1.3.1              </span></span>
<span><span class="co">##  [63] data.table_1.17.8         BiocSingular_1.24.0      </span></span>
<span><span class="co">##  [65] ScaledMatrix_1.16.0       metapod_1.16.0           </span></span>
<span><span class="co">##  [67] XVector_0.48.0            RcppAnnoy_0.0.22         </span></span>
<span><span class="co">##  [69] RANN_2.6.2                ggrepel_0.9.6            </span></span>
<span><span class="co">##  [71] pillar_1.11.0             stringr_1.5.1            </span></span>
<span><span class="co">##  [73] limma_3.64.1              RcppHungarian_0.3        </span></span>
<span><span class="co">##  [75] splines_4.5.1             dplyr_1.1.4              </span></span>
<span><span class="co">##  [77] lattice_0.22-7            bit_4.6.0                </span></span>
<span><span class="co">##  [79] annotate_1.86.1           tidyselect_1.2.1         </span></span>
<span><span class="co">##  [81] locfit_1.5-9.12           Biostrings_2.76.0        </span></span>
<span><span class="co">##  [83] knitr_1.50                gridExtra_2.3            </span></span>
<span><span class="co">##  [85] edgeR_4.6.3               scattermore_1.2          </span></span>
<span><span class="co">##  [87] xfun_0.52                 statmod_1.5.0            </span></span>
<span><span class="co">##  [89] stringi_1.8.7             UCSC.utils_1.4.0         </span></span>
<span><span class="co">##  [91] yaml_2.3.10               evaluate_1.0.4           </span></span>
<span><span class="co">##  [93] codetools_0.2-20          tibble_3.3.0             </span></span>
<span><span class="co">##  [95] BiocManager_1.30.26       graph_1.86.0             </span></span>
<span><span class="co">##  [97] cli_3.6.5                 uwot_0.2.3               </span></span>
<span><span class="co">##  [99] leidenAlg_1.1.5           xtable_1.8-4             </span></span>
<span><span class="co">## [101] systemfonts_1.2.3         jquerylib_0.1.4          </span></span>
<span><span class="co">## [103] Rcpp_1.1.0                png_0.1-8                </span></span>
<span><span class="co">## [105] XML_3.99-0.18             parallel_4.5.1           </span></span>
<span><span class="co">## [107] pkgdown_2.1.3             blob_1.2.4               </span></span>
<span><span class="co">## [109] mclust_6.1.1              ggalluvial_0.12.5        </span></span>
<span><span class="co">## [111] sparseMatrixStats_1.20.0  viridisLite_0.4.2        </span></span>
<span><span class="co">## [113] GSEABase_1.70.0           scales_1.4.0             </span></span>
<span><span class="co">## [115] purrr_1.1.0               crayon_1.5.3             </span></span>
<span><span class="co">## [117] rlang_1.1.6               cowplot_1.2.0            </span></span>
<span><span class="co">## [119] KEGGREST_1.48.1</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="acknowledgments">Acknowledgments<a class="anchor" aria-label="anchor" href="#acknowledgments"></a>
</h2>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-R-rmarkdown" class="csl-entry">
Allaire, JJ, Yihui Xie, Christophe Dervieux, Jonathan McPherson, Javier
Luraschi, Kevin Ushey, Aron Atkins, et al. 2024. <em>Rmarkdown: Dynamic
Documents for r</em>. <a href="https://github.com/rstudio/rmarkdown" class="external-link">https://github.com/rstudio/rmarkdown</a>.
</div>
<div id="ref-amezquita2020orchestrating" class="csl-entry">
Amezquita, Robert A, Aaron TL Lun, Etienne Becht, Vince J Carey, Lindsay
N Carpp, Ludwig Geistlinger, Federico Marini, et al. 2020.
<span>“Orchestrating Single-Cell Analysis with Bioconductor.”</span>
<em>Nature Methods</em> 17 (2): 137–45.
</div>
<div id="ref-aran2019reference" class="csl-entry">
Aran, Dvir, Agnieszka P Looney, Leqian Liu, Esther Wu, Valerie Fong,
Austin Hsu, Suzanna Chak, et al. 2019. <span>“Reference-Based Analysis
of Lung Single-Cell Sequencing Reveals a Transitional Profibrotic
Macrophage.”</span> <em>Nature Immunology</em> 20 (2): 163–72.
</div>
<div id="ref-bhuva2024library" class="csl-entry">
Bhuva, Dharmesh D, Chin Wee Tan, Agus Salim, Claire Marceaux, Marie A
Pickering, Jinjin Chen, Malvika Kharbanda, et al. 2024. <span>“Library
Size Confounds Biology in Spatial Transcriptomics Data.”</span>
<em>Genome Biology</em> 25 (1): 99.
</div>
<div id="ref-fu2024bidcell" class="csl-entry">
Fu, Xiaohang, Yingxin Lin, David M Lin, Daniel Mechtersheimer, Chuhan
Wang, Farhan Ameen, Shila Ghazanfar, Ellis Patrick, Jinman Kim, and Jean
YH Yang. 2024. <span>“BIDCell: Biologically-Informed Self-Supervised
Learning for Segmentation of Subcellular Spatial Transcriptomics
Data.”</span> <em>Nature Communications</em> 15 (1): 509.
</div>
<div id="ref-liu2024hoodscanr" class="csl-entry">
Liu, Ning, Jarryd Martin, Dharmesh D Bhuva, Jinjin Chen, Mengbo Li,
Samuel C Lee, Malvika Kharbanda, et al. 2024. <span>“hoodscanR:
Profiling Single-Cell Neighborhoods in Spatial Transcriptomics
Data.”</span> <em>bioRxiv</em>, 2024–03.
</div>
<div id="ref-R-prettydoc" class="csl-entry">
Qiu, Yixuan. 2021. <em>Prettydoc: Creating Pretty Documents from r
Markdown</em>. <a href="https://github.com/yixuan/prettydoc" class="external-link">https://github.com/yixuan/prettydoc</a>.
</div>
<div id="ref-salim2024spanorm" class="csl-entry">
Salim, Agus, Dharmesh D Bhuva, Carissa Chen, Pengyi Yang, Melissa J
Davis, and Jean YH Yang. 2024. <span>“SpaNorm: Spatially-Aware
Normalisation for Spatial Transcriptomics Data.”</span>
<em>bioRxiv</em>, 2024–05.
</div>
<div id="ref-singhal2024banksy" class="csl-entry">
Singhal, Vipul, Nigel Chou, Joseph Lee, Yifei Yue, Jinyue Liu, Wan Kee
Chock, Li Lin, et al. 2024. <span>“BANKSY Unifies Cell Typing and Tissue
Domain Segmentation for Scalable Spatial Omics Data Analysis.”</span>
<em>Nature Genetics</em> 56 (3): 431–41.
</div>
<div id="ref-totty2024spotsweeper" class="csl-entry">
Totty, Michael, Stephanie C Hicks, and Boyi Guo. 2024.
<span>“SpotSweeper: Spatially-Aware Quality Control for Spatial
Transcriptomics.”</span> <em>bioRxiv</em>.
</div>
<div id="ref-R-ggplot2" class="csl-entry">
Wickham, Hadley, Winston Chang, Lionel Henry, Thomas Lin Pedersen,
Kohske Takahashi, Claus Wilke, Kara Woo, Hiroaki Yutani, Dewey
Dunnington, and Teun van den Brand. 2024. <em>Ggplot2: Create Elegant
Data Visualisations Using the Grammar of Graphics</em>. <a href="https://ggplot2.tidyverse.org" class="external-link">https://ggplot2.tidyverse.org</a>.
</div>
<div id="ref-R-knitr" class="csl-entry">
Xie, Yihui. 2024. <em>Knitr: A General-Purpose Package for Dynamic
Report Generation in r</em>. <a href="https://yihui.org/knitr/" class="external-link">https://yihui.org/knitr/</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Dharmesh D. Bhuva, Chin Wee Tan.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
